{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The goal of this analysis is to select and predict CRISPR perturbations of desired motifs at a specific loci.\n",
    "\n",
    "# Computational setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-11-12 13:47:07.676473: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-11-12 13:47:07.676506: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-11-12 13:47:07.678170: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-11-12 13:47:07.686425: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-11-12 13:47:08.941451: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘fasta’: File exists\n",
      "mkdir: cannot create directory ‘json/crispr’: File exists\n",
      "mkdir: cannot create directory ‘shap/crispr’: File exists\n",
      "mkdir: cannot create directory ‘pisa’: File exists\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tensorflow.python.util import deprecation\n",
    "deprecation._PRINT_DEPRECATION_WARNINGS = False\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from pybedtools import BedTool\n",
    "import keras.backend as K\n",
    "from keras.models import load_model\n",
    "import json\n",
    "\n",
    "## Working options\n",
    "os.chdir(f'/n/projects/mw2098/publications/2024_weilert_acc/code/2_analysis/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "bpreveal_path = '/n/projects/mw2098/publications/2024_weilert_acc/public/software/bpreveal_404/'\n",
    "python_path = '/home/mw2098/anaconda3/envs/bpreveal_404/bin/python'\n",
    "\n",
    "sys.path.insert(0, f'{bpreveal_path}/src')\n",
    "import losses\n",
    "\n",
    "## Custom functions\n",
    "sys.path.insert(0, f'scripts/py/functions')\n",
    "from functional import shuffle_seqs, one_hot_encode_sequence, one_hot_encode_sequences, \\\n",
    "    one_hot_decode_sequence, insert_motif, logitsToProfile\n",
    "from motifs import extract_seqs_from_df, resize_coordinates\n",
    "\n",
    "#Pre-existing variables\n",
    "region_dict = {'oct4': 'narrowpeak/mesc_oct4_nexus_peaks.narrowPeak',\n",
    "               'sox2': 'narrowpeak/mesc_sox2_nexus_peaks.narrowPeak',\n",
    "               'klf4': 'narrowpeak/mesc_klf4_nexus_peaks.narrowPeak',\n",
    "               'nanog': 'narrowpeak/mesc_nanog_nexus_peaks.narrowPeak',\n",
    "               'zic3': 'narrowpeak/mesc_zic3_nexus_peaks.narrowPeak'}\n",
    "concentration_atac_timepoints = list(range(0, 16, 3))\n",
    "modeling_design_dict = {\n",
    "    'bpnet_osknz_fold1':{\n",
    "        'tasks': list(region_dict.keys()),\n",
    "        'cov': {k: {'pos': 'bw/mesc_' + k + '_nexus_combined_positive.bw',\n",
    "                    'neg': 'bw/mesc_' + k + '_nexus_combined_negative.bw'}\n",
    "                for k in region_dict.keys()},\n",
    "        'model_dir': 'models/bpnet_osknz_fold1.model',\n",
    "        'num-channels': 2\n",
    "    },\n",
    "    'bpnet_osknz_fold2':{\n",
    "        'tasks': list(region_dict.keys()),\n",
    "        'cov': {k: {'pos': 'bw/mesc_' + k + '_nexus_combined_positive.bw',\n",
    "                    'neg': 'bw/mesc_' + k + '_nexus_combined_negative.bw'}\n",
    "                for k in region_dict.keys()},\n",
    "        'model_dir': 'models/bpnet_osknz_fold2.model',\n",
    "        'num-channels': 2\n",
    "    },\n",
    "    'bpnet_osknz_fold3':{\n",
    "        'tasks': list(region_dict.keys()),\n",
    "        'cov': {k: {'pos': 'bw/mesc_' + k + '_nexus_combined_positive.bw',\n",
    "                    'neg': 'bw/mesc_' + k + '_nexus_combined_negative.bw'}\n",
    "                for k in region_dict.keys()},\n",
    "        'model_dir': 'models/bpnet_osknz_fold3.model',\n",
    "        'num-channels': 2\n",
    "    },\n",
    "    'atac_wt_fold1':{\n",
    "        'tasks': ['atac'],\n",
    "        'cov': 'bw/mesc_native_atac_cutsites_combined.bw',\n",
    "        'model_dir': 'models/atac_wt_fold1_residual.model/',\n",
    "        'num-channels': 1\n",
    "    },\n",
    "    'atac_wt_fold2':{\n",
    "        'tasks': ['atac'],\n",
    "        'cov': 'bw/mesc_native_atac_cutsites_combined.bw',\n",
    "        'model_dir': 'models/atac_wt_fold2_residual.model/',\n",
    "        'num-channels': 1\n",
    "    },\n",
    "    'atac_wt_fold3':{\n",
    "        'tasks': ['atac'],\n",
    "        'cov': 'bw/mesc_native_atac_cutsites_combined.bw',\n",
    "        'model_dir': 'models/atac_wt_fold3_residual.model/',\n",
    "        'num-channels': 1\n",
    "    }\n",
    "}\n",
    "for timepoint in concentration_atac_timepoints:\n",
    "    modeling_design_dict[f'atac_{timepoint}h_fold1'] = {\n",
    "            'tasks': ['atac'],\n",
    "            'cov': f'bw/GSE174774_mesc_atac_{timepoint}h_combined.bw',\n",
    "            'model_dir': f'models/atac_{timepoint}h_fold1_residual.model/',\n",
    "            'num-channels': 1\n",
    "    }\n",
    "\n",
    "input_length = 2032\n",
    "output_length = 1000\n",
    "flank_length = (input_length - output_length)//2\n",
    "trials = 256\n",
    "seed = 2356\n",
    "\n",
    "genome = '../0_setup/fa/mm10.fa'\n",
    "figure_path = 'figures/13_perturb_crispr'\n",
    "regions_path = 'bed/mapped_motifs/all_islands_curated_0based_sized_to_input.bed'\n",
    "motifs_path = 'tsv/mapped_motifs/all_instances_curated_0based_w_perturb.tsv.gz'\n",
    "\n",
    "working_dir = os.getcwd()\n",
    "gpu_header = ['#!/usr/bin/bash',\n",
    "              '#SBATCH --job-name bpnet_training',\n",
    "              '#SBATCH --output=slurm_%j.log',\n",
    "              '#SBATCH --mem=200gb', '#SBATCH --time=72:00:00',\n",
    "              '#SBATCH --partition=gpu', '#SBATCH --gres=gpu:a100:1',\n",
    "              '#SBATCH --cpus-per-task=10',\n",
    "              'source /home/mw2098/.bashrc',\n",
    "              'conda deactivate',\n",
    "              'conda activate bpreveal_404',\n",
    "              f'cd {working_dir}']\n",
    "\n",
    "!mkdir -p {figure_path} tsv/genomic/crispr\n",
    "!mkdir fasta\n",
    "!mkdir json/crispr\n",
    "!mkdir shap/crispr\n",
    "!mkdir pisa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create SLURM array job function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_slurm_array(cmds, output_file, simultaneous_jobs = 4):\n",
    "    total_jobs = len(cmds)\n",
    "    array_header = ['#!/usr/bin/bash',\n",
    "                  '#SBATCH --job-name bpnet_training',\n",
    "                  '#SBATCH --output=slurm_%j.log',\n",
    "                  f'#SBATCH --array=1-{total_jobs}%{simultaneous_jobs}',\n",
    "                  '#SBATCH --mem=200gb', \n",
    "                  '#SBATCH --time=72:00:00',\n",
    "                  '#SBATCH --partition=gpu', \n",
    "                  '#SBATCH --gres=gpu:a100:1',\n",
    "                  '#SBATCH --cpus-per-task=10',\n",
    "                  'source /home/mw2098/.bashrc',\n",
    "                  'conda deactivate',\n",
    "                  'conda activate bpreveal_404',\n",
    "                  f'cd {working_dir}']\n",
    "    # array_cmds = array_header\n",
    "    for i,cmd in enumerate(cmds):\n",
    "        array_cmd = [\"if [[ ${{SLURM_ARRAY_TASK_ID}} == {0:d} ]] ; then\\n\".format(i+1), \n",
    "                    \"    {0:s}\\n\".format(cmd),\n",
    "                    \"fi\\n\\n\"]\n",
    "        array_header = array_header + array_cmd \n",
    "        \n",
    "    with open(output_file, mode='wt') as slurm:\n",
    "        slurm.write('\\n'.join(array_header))\n",
    "        slurm.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import regions of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_df = pd.read_csv(regions_path,  sep = '\\t', names = ['chrom','start','end','region_id','score','strand'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perturb Avsec et al published site\n",
    "\n",
    "This is the site CRISPRed in Figure 6 of Avsec et al (2021). What was not previously noted was that there is a low-affinity Oct4-Sox2 motif between the Sox2 and Nanog motifs that is highly contributing to accessibility and binding. This is an ideal candidate to explore (1) can low-affinity motifs be cooperative, (2) by increasing affinity, can we increase cooperativity and (3) how does the distance of the two motifs influence the designated cooperativity?\n",
    "\n",
    "For this case, motifA will be Sox2, and motifB will be Oct4-Sox2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       motif     wt_seq  motif_window_start  motif_window_end  pattern_center  \\\n",
      "209969  Sox2  GCCTTTGTT                 961               970             965   \n",
      "\n",
      "        seq_match_quantile  \n",
      "209969             0.96575  \n",
      "            motif           wt_seq  motif_window_start  motif_window_end  \\\n",
      "207188  Oct4-Sox2  AATTATAATGATAAT                 995              1010   \n",
      "\n",
      "        pattern_center  seq_match_quantile  \n",
      "207188            1002            0.605285  \n"
     ]
    }
   ],
   "source": [
    "sox2_motif_id = 283919\n",
    "oct4sox2_motif_id = 139897\n",
    "region_id = 97554\n",
    "\n",
    "#Isolate motifs\n",
    "motifs_df = pd.read_csv(motifs_path,  sep = '\\t')\n",
    "motifs_df['pattern_center'] = motifs_df['pattern_center'] + flank_length\n",
    "sox2_motif_df = motifs_df[motifs_df.motif_id==sox2_motif_id]\n",
    "sox2_motif_df['wt_seq'] = extract_seqs_from_df(sox2_motif_df, genome)\n",
    "oct4sox2_motif_df = motifs_df[motifs_df.motif_id==oct4sox2_motif_id]\n",
    "oct4sox2_motif_df['wt_seq'] = extract_seqs_from_df(oct4sox2_motif_df, genome)\n",
    "\n",
    "#Keep metadata of use\n",
    "columns_of_interest = ['motif', 'wt_seq', 'motif_window_start', 'motif_window_end', 'pattern_center', 'seq_match_quantile']\n",
    "sox2_df = sox2_motif_df[columns_of_interest]\n",
    "oct4sox2_df = oct4sox2_motif_df[columns_of_interest]\n",
    "print(sox2_df)\n",
    "print(oct4sox2_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# region_start = sox2_motif_df.region_start.values[0]\n",
    "# region_start\n",
    "# #original: 85539378"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Print region start coordinate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "85538671"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_start = sox2_motif_df.region_start.values[0]\n",
    "region_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overwrite distance injection for `dist_coop` scenario."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current center-to-center distance:  [37]\n"
     ]
    }
   ],
   "source": [
    "print('Current center-to-center distance: ', oct4sox2_df.pattern_center.values - sox2_df.pattern_center.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distal_center_to_center_distance = 140\n",
    "distal_to_add = distal_center_to_center_distance - (oct4sox2_df.pattern_center.values - sox2_df.pattern_center.values)[0]\n",
    "distal_to_add"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Sox2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define all CRISPR scenarios for Sox2. Sox2 will not change in position or be enhanced by affinity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq</th>\n",
       "      <th>motif</th>\n",
       "      <th>wt_seq</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "      <th>pattern_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>null</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  state forward_seq motif     wt_seq  motif_window_start  motif_window_end  \\\n",
       "0    AB   GCCTTTGTT  Sox2  GCCTTTGTT                 961               970   \n",
       "1     A   GCCTTTGTT  Sox2  GCCTTTGTT                 961               970   \n",
       "2     B   GCCTAGGTT  Sox2  GCCTTTGTT                 961               970   \n",
       "3  null   GCCTAGGTT  Sox2  GCCTTTGTT                 961               970   \n",
       "\n",
       "   pattern_center  \n",
       "0             965  \n",
       "1             965  \n",
       "2             965  \n",
       "3             965  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sox2_states_df = pd.DataFrame([['AB', 'A', 'B', 'null'], ['GCCTTTGTT', 'GCCTTTGTT', 'GCCTAGGTT', 'GCCTAGGTT']])\\\n",
    "    .transpose()\\\n",
    "    .rename({0: 'state', 1:'forward_seq'}, axis = 1)\n",
    "sox2_states_df['motif'] = 'Sox2'\n",
    "sox2_states_df = sox2_states_df.merge(sox2_df.drop(['seq_match_quantile'], axis = 1))\n",
    "sox2_states_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Oct4-Sox2\n",
    "\n",
    "First, define default state where the low-affinity WT Oct4-Sox2 is mutated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "oct4sox2_states_df = pd.DataFrame([['WT_coop', 'WT_coop', 'WT_coop', 'WT_coop', \n",
    "                                    'enh_coop', 'enh_coop', 'enh_coop', 'enh_coop', \n",
    "                                    'dist_coop', 'dist_coop', 'dist_coop', 'dist_coop'],\n",
    "                                   ['AB','A','B','null', \n",
    "                                    'AB','A','B','null',\n",
    "                                    'AB','A','B','null'], \n",
    "                                   ['AATTATAATGATAAT', 'AATCATAAGGATAAT', 'AATTATAATGATAAT', 'AATCATAAGGATAAT',\n",
    "                                    'AATTGTAATGCTAAT', 'AATCATAAGGATAAT', 'AATTGTAATGCTAAT', 'AATCATAAGGATAAT',\n",
    "                                    'AATTATAATGATAAT', 'AATCATAAGGATAAT', 'AATTATAATGATAAT', 'AATCATAAGGATAAT'],\n",
    "                                   [0, 0, 0, 0, 0, 0, 0, 0, distal_to_add, distal_to_add, distal_to_add, distal_to_add]])\\\n",
    "    .transpose()\\\n",
    "    .rename({0: 'scenario', 1:'state', 2:'forward_seq', 3: 'distance_to_add'}, axis = 1)\n",
    "oct4sox2_states_df['motif'] = 'Oct4-Sox2'\n",
    "oct4sox2_states_df = oct4sox2_states_df.merge(oct4sox2_df.drop(['seq_match_quantile'], axis = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "oct4sox2_states_df['motif_window_start'] = oct4sox2_states_df['motif_window_start'] + oct4sox2_states_df['distance_to_add']\n",
    "oct4sox2_states_df['motif_window_end'] = oct4sox2_states_df['motif_window_end'] + oct4sox2_states_df['distance_to_add']\n",
    "oct4sox2_states_df['pattern_center'] = oct4sox2_states_df['pattern_center'] + oct4sox2_states_df['distance_to_add']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq</th>\n",
       "      <th>distance_to_add</th>\n",
       "      <th>motif</th>\n",
       "      <th>wt_seq</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "      <th>pattern_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     scenario state      forward_seq distance_to_add      motif  \\\n",
       "0     WT_coop    AB  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "1     WT_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "2     WT_coop     B  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "3     WT_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "4    enh_coop    AB  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "5    enh_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "6    enh_coop     B  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "7    enh_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "8   dist_coop    AB  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "9   dist_coop     A  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "10  dist_coop     B  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "11  dist_coop  null  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "\n",
       "             wt_seq motif_window_start motif_window_end pattern_center  \n",
       "0   AATTATAATGATAAT                995             1010           1002  \n",
       "1   AATTATAATGATAAT                995             1010           1002  \n",
       "2   AATTATAATGATAAT                995             1010           1002  \n",
       "3   AATTATAATGATAAT                995             1010           1002  \n",
       "4   AATTATAATGATAAT                995             1010           1002  \n",
       "5   AATTATAATGATAAT                995             1010           1002  \n",
       "6   AATTATAATGATAAT                995             1010           1002  \n",
       "7   AATTATAATGATAAT                995             1010           1002  \n",
       "8   AATTATAATGATAAT               1098             1113           1105  \n",
       "9   AATTATAATGATAAT               1098             1113           1105  \n",
       "10  AATTATAATGATAAT               1098             1113           1105  \n",
       "11  AATTATAATGATAAT               1098             1113           1105  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oct4sox2_states_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge coordinates together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_scenarios_df = oct4sox2_states_df.merge(sox2_states_df, how = 'left', on = 'state', suffixes=('_OS', '_S'))\n",
    "crispr_scenarios_df[['genomic_start_OS', 'genomic_end_OS', 'genomic_center_OS']] = crispr_scenarios_df[['motif_window_start_OS', 'motif_window_end_OS', 'pattern_center_OS']] + region_start\n",
    "crispr_scenarios_df[['genomic_start_S', 'genomic_end_S', 'genomic_center_S']] = crispr_scenarios_df[['motif_window_start_S', 'motif_window_end_S', 'pattern_center_S']] + region_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq_OS</th>\n",
       "      <th>distance_to_add</th>\n",
       "      <th>motif_OS</th>\n",
       "      <th>wt_seq_OS</th>\n",
       "      <th>motif_window_start_OS</th>\n",
       "      <th>motif_window_end_OS</th>\n",
       "      <th>pattern_center_OS</th>\n",
       "      <th>forward_seq_S</th>\n",
       "      <th>motif_S</th>\n",
       "      <th>wt_seq_S</th>\n",
       "      <th>motif_window_start_S</th>\n",
       "      <th>motif_window_end_S</th>\n",
       "      <th>pattern_center_S</th>\n",
       "      <th>genomic_start_OS</th>\n",
       "      <th>genomic_end_OS</th>\n",
       "      <th>genomic_center_OS</th>\n",
       "      <th>genomic_start_S</th>\n",
       "      <th>genomic_end_S</th>\n",
       "      <th>genomic_center_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     scenario state   forward_seq_OS distance_to_add   motif_OS  \\\n",
       "0     WT_coop    AB  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "1     WT_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "2     WT_coop     B  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "3     WT_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "4    enh_coop    AB  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "5    enh_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "6    enh_coop     B  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "7    enh_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "8   dist_coop    AB  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "9   dist_coop     A  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "10  dist_coop     B  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "11  dist_coop  null  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "\n",
       "          wt_seq_OS motif_window_start_OS motif_window_end_OS  \\\n",
       "0   AATTATAATGATAAT                   995                1010   \n",
       "1   AATTATAATGATAAT                   995                1010   \n",
       "2   AATTATAATGATAAT                   995                1010   \n",
       "3   AATTATAATGATAAT                   995                1010   \n",
       "4   AATTATAATGATAAT                   995                1010   \n",
       "5   AATTATAATGATAAT                   995                1010   \n",
       "6   AATTATAATGATAAT                   995                1010   \n",
       "7   AATTATAATGATAAT                   995                1010   \n",
       "8   AATTATAATGATAAT                  1098                1113   \n",
       "9   AATTATAATGATAAT                  1098                1113   \n",
       "10  AATTATAATGATAAT                  1098                1113   \n",
       "11  AATTATAATGATAAT                  1098                1113   \n",
       "\n",
       "   pattern_center_OS forward_seq_S motif_S   wt_seq_S  motif_window_start_S  \\\n",
       "0               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "1               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "2               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "3               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "4               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "5               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "6               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "7               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "8               1105     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "9               1105     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "10              1105     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "11              1105     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "\n",
       "    motif_window_end_S  pattern_center_S genomic_start_OS genomic_end_OS  \\\n",
       "0                  970               965         85539666       85539681   \n",
       "1                  970               965         85539666       85539681   \n",
       "2                  970               965         85539666       85539681   \n",
       "3                  970               965         85539666       85539681   \n",
       "4                  970               965         85539666       85539681   \n",
       "5                  970               965         85539666       85539681   \n",
       "6                  970               965         85539666       85539681   \n",
       "7                  970               965         85539666       85539681   \n",
       "8                  970               965         85539769       85539784   \n",
       "9                  970               965         85539769       85539784   \n",
       "10                 970               965         85539769       85539784   \n",
       "11                 970               965         85539769       85539784   \n",
       "\n",
       "   genomic_center_OS  genomic_start_S  genomic_end_S  genomic_center_S  \n",
       "0           85539673         85539632       85539641          85539636  \n",
       "1           85539673         85539632       85539641          85539636  \n",
       "2           85539673         85539632       85539641          85539636  \n",
       "3           85539673         85539632       85539641          85539636  \n",
       "4           85539673         85539632       85539641          85539636  \n",
       "5           85539673         85539632       85539641          85539636  \n",
       "6           85539673         85539632       85539641          85539636  \n",
       "7           85539673         85539632       85539641          85539636  \n",
       "8           85539776         85539632       85539641          85539636  \n",
       "9           85539776         85539632       85539641          85539636  \n",
       "10          85539776         85539632       85539641          85539636  \n",
       "11          85539776         85539632       85539641          85539636  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_scenarios_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate predictions based on the different scenarios\n",
    "\n",
    "First, prepare WT genomic sequences according to input accessibility features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_region_df = regions_df[regions_df.region_id==region_id]\n",
    "crispr_region_df['wt_seq'] = extract_seqs_from_df(crispr_region_df, genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       chrom     start       end  region_id  score strand  \\\n",
      "97547  chr10  85538671  85540703      97547      0      .   \n",
      "\n",
      "                                                  wt_seq  \n",
      "97547  TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...  \n",
      "TCCTACAGAC\n",
      "TATCAAACAC\n"
     ]
    }
   ],
   "source": [
    "print(crispr_region_df)\n",
    "print(crispr_region_df.wt_seq.values[0][:10])\n",
    "print(crispr_region_df.wt_seq.values[0][-10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_region_df.to_csv('tsv/genomic/crispr/crispr_coop_coord.bed', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_model = load_model(modeling_design_dict['atac_wt_fold1']['model_dir'], custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, inject motifs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_seq = crispr_region_df.wt_seq.values[0]\n",
    "\n",
    "crispr_seqs = []\n",
    "for i,row in crispr_scenarios_df.iterrows():\n",
    "    if row.scenario=='dist_coop':\n",
    "        original_os_start = crispr_scenarios_df.motif_window_start_OS.values[0]\n",
    "        original_os_end = crispr_scenarios_df.motif_window_end_OS.values[0]\n",
    "        new_os_start = row.motif_window_start_OS\n",
    "        new_os_end = row.motif_window_end_OS\n",
    "        \n",
    "        #Overwrite Sox2 motif\n",
    "        s_seq = wt_seq[:(row.motif_window_start_S)] + row.forward_seq_S + wt_seq[(row.motif_window_end_S):]\n",
    "        \n",
    "        #Overwrite Oct4-Sox2 motif by flipping the sequence\n",
    "        final_seq = s_seq[:(original_os_start)] + \\\n",
    "        wt_seq[(new_os_start):(new_os_end)] + \\\n",
    "        s_seq[(original_os_end):(new_os_start)] + \\\n",
    "        row.forward_seq_OS + \\\n",
    "        s_seq[(new_os_end):]\n",
    "        \n",
    "        assert len(final_seq)==input_length\n",
    "        crispr_seqs.append(final_seq)\n",
    "    else:\n",
    "        #Overwrite Sox2 motif\n",
    "        s_seq = wt_seq[:(row.motif_window_start_S)] + row.forward_seq_S + wt_seq[(row.motif_window_end_S):]\n",
    "        \n",
    "        #Overwrite Oct4-Sox2 motif\n",
    "        final_seq = s_seq[:(row.motif_window_start_OS)] + row.forward_seq_OS + s_seq[(row.motif_window_end_OS):]\n",
    "        assert len(final_seq)==input_length\n",
    "        crispr_seqs.append(final_seq)\n",
    "\n",
    "crispr_scenarios_df['inj_seq'] = crispr_seqs\n",
    "crispr_scenarios_df['scenario_index'] = list(range(crispr_scenarios_df.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq_OS</th>\n",
       "      <th>distance_to_add</th>\n",
       "      <th>motif_OS</th>\n",
       "      <th>wt_seq_OS</th>\n",
       "      <th>motif_window_start_OS</th>\n",
       "      <th>motif_window_end_OS</th>\n",
       "      <th>pattern_center_OS</th>\n",
       "      <th>forward_seq_S</th>\n",
       "      <th>motif_S</th>\n",
       "      <th>wt_seq_S</th>\n",
       "      <th>motif_window_start_S</th>\n",
       "      <th>motif_window_end_S</th>\n",
       "      <th>pattern_center_S</th>\n",
       "      <th>genomic_start_OS</th>\n",
       "      <th>genomic_end_OS</th>\n",
       "      <th>genomic_center_OS</th>\n",
       "      <th>genomic_start_S</th>\n",
       "      <th>genomic_end_S</th>\n",
       "      <th>genomic_center_S</th>\n",
       "      <th>inj_seq</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>inj_seq_85539400_to_85540000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>1</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>2</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>3</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>4</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>5</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTGTAATGCTAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>6</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>enh_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>0</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>995</td>\n",
       "      <td>1010</td>\n",
       "      <td>1002</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539666</td>\n",
       "      <td>85539681</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>7</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>8</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>A</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>9</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>B</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>10</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>dist_coop</td>\n",
       "      <td>null</td>\n",
       "      <td>AATCATAAGGATAAT</td>\n",
       "      <td>103</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>AATTATAATGATAAT</td>\n",
       "      <td>1098</td>\n",
       "      <td>1113</td>\n",
       "      <td>1105</td>\n",
       "      <td>GCCTAGGTT</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>GCCTTTGTT</td>\n",
       "      <td>961</td>\n",
       "      <td>970</td>\n",
       "      <td>965</td>\n",
       "      <td>85539769</td>\n",
       "      <td>85539784</td>\n",
       "      <td>85539776</td>\n",
       "      <td>85539632</td>\n",
       "      <td>85539641</td>\n",
       "      <td>85539636</td>\n",
       "      <td>TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...</td>\n",
       "      <td>11</td>\n",
       "      <td>TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     scenario state   forward_seq_OS distance_to_add   motif_OS  \\\n",
       "0     WT_coop    AB  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "1     WT_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "2     WT_coop     B  AATTATAATGATAAT               0  Oct4-Sox2   \n",
       "3     WT_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "4    enh_coop    AB  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "5    enh_coop     A  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "6    enh_coop     B  AATTGTAATGCTAAT               0  Oct4-Sox2   \n",
       "7    enh_coop  null  AATCATAAGGATAAT               0  Oct4-Sox2   \n",
       "8   dist_coop    AB  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "9   dist_coop     A  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "10  dist_coop     B  AATTATAATGATAAT             103  Oct4-Sox2   \n",
       "11  dist_coop  null  AATCATAAGGATAAT             103  Oct4-Sox2   \n",
       "\n",
       "          wt_seq_OS motif_window_start_OS motif_window_end_OS  \\\n",
       "0   AATTATAATGATAAT                   995                1010   \n",
       "1   AATTATAATGATAAT                   995                1010   \n",
       "2   AATTATAATGATAAT                   995                1010   \n",
       "3   AATTATAATGATAAT                   995                1010   \n",
       "4   AATTATAATGATAAT                   995                1010   \n",
       "5   AATTATAATGATAAT                   995                1010   \n",
       "6   AATTATAATGATAAT                   995                1010   \n",
       "7   AATTATAATGATAAT                   995                1010   \n",
       "8   AATTATAATGATAAT                  1098                1113   \n",
       "9   AATTATAATGATAAT                  1098                1113   \n",
       "10  AATTATAATGATAAT                  1098                1113   \n",
       "11  AATTATAATGATAAT                  1098                1113   \n",
       "\n",
       "   pattern_center_OS forward_seq_S motif_S   wt_seq_S  motif_window_start_S  \\\n",
       "0               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "1               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "2               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "3               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "4               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "5               1002     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "6               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "7               1002     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "8               1105     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "9               1105     GCCTTTGTT    Sox2  GCCTTTGTT                   961   \n",
       "10              1105     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "11              1105     GCCTAGGTT    Sox2  GCCTTTGTT                   961   \n",
       "\n",
       "    motif_window_end_S  pattern_center_S genomic_start_OS genomic_end_OS  \\\n",
       "0                  970               965         85539666       85539681   \n",
       "1                  970               965         85539666       85539681   \n",
       "2                  970               965         85539666       85539681   \n",
       "3                  970               965         85539666       85539681   \n",
       "4                  970               965         85539666       85539681   \n",
       "5                  970               965         85539666       85539681   \n",
       "6                  970               965         85539666       85539681   \n",
       "7                  970               965         85539666       85539681   \n",
       "8                  970               965         85539769       85539784   \n",
       "9                  970               965         85539769       85539784   \n",
       "10                 970               965         85539769       85539784   \n",
       "11                 970               965         85539769       85539784   \n",
       "\n",
       "   genomic_center_OS  genomic_start_S  genomic_end_S  genomic_center_S  \\\n",
       "0           85539673         85539632       85539641          85539636   \n",
       "1           85539673         85539632       85539641          85539636   \n",
       "2           85539673         85539632       85539641          85539636   \n",
       "3           85539673         85539632       85539641          85539636   \n",
       "4           85539673         85539632       85539641          85539636   \n",
       "5           85539673         85539632       85539641          85539636   \n",
       "6           85539673         85539632       85539641          85539636   \n",
       "7           85539673         85539632       85539641          85539636   \n",
       "8           85539776         85539632       85539641          85539636   \n",
       "9           85539776         85539632       85539641          85539636   \n",
       "10          85539776         85539632       85539641          85539636   \n",
       "11          85539776         85539632       85539641          85539636   \n",
       "\n",
       "                                              inj_seq  scenario_index  \\\n",
       "0   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               0   \n",
       "1   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               1   \n",
       "2   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               2   \n",
       "3   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               3   \n",
       "4   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               4   \n",
       "5   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               5   \n",
       "6   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               6   \n",
       "7   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               7   \n",
       "8   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               8   \n",
       "9   TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...               9   \n",
       "10  TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...              10   \n",
       "11  TCCTACAGACCCCTTTGTTGTCCCACTCTTTAATAGAAATACCTGA...              11   \n",
       "\n",
       "                         inj_seq_85539400_to_85540000  \n",
       "0   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "1   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "2   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "3   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "4   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "5   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "6   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "7   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "8   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "9   TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "10  TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  \n",
       "11  TTTGTTAGACCAGTGGAAGTGGGGATAGAGGTGGGAAGAGAGGATG...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_scenarios_df['inj_seq_85539400_to_85540000'] = [s[(85539400 - crispr_region_df.start.values[0]):(85539400 - crispr_region_df.start.values[0] + 600) ]\n",
    "                                                       for s in crispr_scenarios_df['inj_seq'].values]\n",
    "crispr_scenarios_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_scenarios_df.to_csv('tsv/genomic/crispr/crispr_coop_scenarios.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AATTATAATGATAAT',\n",
       " 'AATCATAAGGATAAT',\n",
       " 'AATTATAATGATAAT',\n",
       " 'AATCATAAGGATAAT',\n",
       " 'AATTGTAATGCTAAT',\n",
       " 'AATCATAAGGATAAT',\n",
       " 'AATTGTAATGCTAAT',\n",
       " 'AATCATAAGGATAAT',\n",
       " 'TGGAAGTTCTCCATT',\n",
       " 'TGGAAGTTCTCCATT',\n",
       " 'TGGAAGTTCTCCATT',\n",
       " 'TGGAAGTTCTCCATT']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[s[(original_os_start):(original_os_end)] for s in crispr_scenarios_df.inj_seq.values]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict injected sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 423ms/step\n",
      "1/1 [==============================] - 0s 307ms/step\n",
      "1/1 [==============================] - 0s 304ms/step\n",
      "1/1 [==============================] - 0s 285ms/step\n",
      "WARNING:tensorflow:5 out of the last 5 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7ff8a68de7a0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has reduce_retracing=True option that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:5 out of the last 5 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7ff8a68de7a0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has reduce_retracing=True option that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 301ms/step\n",
      "WARNING:tensorflow:6 out of the last 6 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7ff8a6c418a0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has reduce_retracing=True option that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:6 out of the last 6 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7ff8a6c418a0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has reduce_retracing=True option that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 286ms/step\n",
      "1/1 [==============================] - 0s 271ms/step\n",
      "1/1 [==============================] - 0s 285ms/step\n",
      "1/1 [==============================] - 0s 311ms/step\n",
      "1/1 [==============================] - 0s 310ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>channel</th>\n",
       "      <th>pred</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>task</th>\n",
       "      <th>model_name</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>genomic_center_OS</th>\n",
       "      <th>genomic_center_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.646414</td>\n",
       "      <td>85539187</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.069548</td>\n",
       "      <td>85539188</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.971074</td>\n",
       "      <td>85539189</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.847520</td>\n",
       "      <td>85539190</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.921221</td>\n",
       "      <td>85539191</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.897421</td>\n",
       "      <td>85539192</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.587497</td>\n",
       "      <td>85539193</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.853134</td>\n",
       "      <td>85539194</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.991755</td>\n",
       "      <td>85539195</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.972739</td>\n",
       "      <td>85539196</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position channel      pred  genomic_position  task         model_name  \\\n",
       "0         0       0  0.646414          85539187  oct4  bpnet_osknz_fold1   \n",
       "1         1       0  1.069548          85539188  oct4  bpnet_osknz_fold1   \n",
       "2         2       0  0.971074          85539189  oct4  bpnet_osknz_fold1   \n",
       "3         3       0  0.847520          85539190  oct4  bpnet_osknz_fold1   \n",
       "4         4       0  0.921221          85539191  oct4  bpnet_osknz_fold1   \n",
       "5         5       0  0.897421          85539192  oct4  bpnet_osknz_fold1   \n",
       "6         6       0  0.587497          85539193  oct4  bpnet_osknz_fold1   \n",
       "7         7       0  0.853134          85539194  oct4  bpnet_osknz_fold1   \n",
       "8         8       0  0.991755          85539195  oct4  bpnet_osknz_fold1   \n",
       "9         9       0  0.972739          85539196  oct4  bpnet_osknz_fold1   \n",
       "\n",
       "   scenario_index scenario state genomic_center_OS  genomic_center_S  \n",
       "0               0  WT_coop    AB          85539673          85539636  \n",
       "1               0  WT_coop    AB          85539673          85539636  \n",
       "2               0  WT_coop    AB          85539673          85539636  \n",
       "3               0  WT_coop    AB          85539673          85539636  \n",
       "4               0  WT_coop    AB          85539673          85539636  \n",
       "5               0  WT_coop    AB          85539673          85539636  \n",
       "6               0  WT_coop    AB          85539673          85539636  \n",
       "7               0  WT_coop    AB          85539673          85539636  \n",
       "8               0  WT_coop    AB          85539673          85539636  \n",
       "9               0  WT_coop    AB          85539673          85539636  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Convert logits and logcounts to human-readable ChIP-nexus profile with counts\n",
    "crispr_df = pd.DataFrame()\n",
    "models_of_interest = ['bpnet_osknz_fold1', 'atac_wt_fold1', 'atac_wt_fold2', 'atac_wt_fold3'] + \\\n",
    "[f'atac_{timepoint}h_fold1' for timepoint in concentration_atac_timepoints]\n",
    "for model_name in models_of_interest:\n",
    "\n",
    "    acc_model = load_model(modeling_design_dict[model_name]['model_dir'], \n",
    "                           custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})\n",
    "    crispr_preds = acc_model.predict(one_hot_encode_sequences(crispr_scenarios_df['inj_seq'].values))\n",
    "\n",
    "    tasks = modeling_design_dict[model_name]['tasks']\n",
    "    tasks_n = len(tasks)\n",
    "    \n",
    "    for j,task in enumerate(tasks):\n",
    "        for i in crispr_scenarios_df['scenario_index'].values:\n",
    "            profile = logitsToProfile(logitsAcrossSingleRegion = crispr_preds[j][i], \n",
    "                                      logCountsAcrossSingleRegion = crispr_preds[j+tasks_n][i])\n",
    "            #Convert to tidy pd.df\n",
    "            df = pd.DataFrame(profile, columns = list(range(modeling_design_dict[model_name]['num-channels'])))\n",
    "            df['position'] = list(range(df.shape[0]))\n",
    "            df = df.melt(id_vars = 'position', var_name = 'channel', value_name = 'pred')\n",
    "            df['genomic_position'] = df['position'] + region_start + flank_length\n",
    "            df['task'] = task\n",
    "            df['model_name'] = model_name\n",
    "            df['scenario_index'] = i\n",
    "            crispr_df = pd.concat([crispr_df, df])\n",
    "crispr_df = crispr_df.merge(crispr_scenarios_df[['scenario_index', 'scenario', 'state', 'genomic_center_OS', 'genomic_center_S']], how = 'left')\n",
    "crispr_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_df.to_csv('tsv/genomic/crispr/crispr_coop_predictions.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict biased sequences for comparison with the experimental predictions in a fair manner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 342ms/step\n",
      "1/1 [==============================] - 0s 454ms/step\n",
      "1/1 [==============================] - 0s 465ms/step\n",
      "1/1 [==============================] - 0s 416ms/step\n",
      "1/1 [==============================] - 0s 433ms/step\n",
      "1/1 [==============================] - 0s 409ms/step\n",
      "1/1 [==============================] - 0s 418ms/step\n",
      "1/1 [==============================] - 0s 423ms/step\n",
      "1/1 [==============================] - 0s 429ms/step\n",
      "1/1 [==============================] - 0s 416ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>channel</th>\n",
       "      <th>pred</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>task</th>\n",
       "      <th>model_name</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>genomic_center_OS</th>\n",
       "      <th>genomic_center_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.646414</td>\n",
       "      <td>85539187</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.069548</td>\n",
       "      <td>85539188</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.971074</td>\n",
       "      <td>85539189</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.847520</td>\n",
       "      <td>85539190</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.921221</td>\n",
       "      <td>85539191</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.897421</td>\n",
       "      <td>85539192</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.587497</td>\n",
       "      <td>85539193</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.853134</td>\n",
       "      <td>85539194</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.991755</td>\n",
       "      <td>85539195</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.972739</td>\n",
       "      <td>85539196</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position channel      pred  genomic_position  task         model_name  \\\n",
       "0         0       0  0.646414          85539187  oct4  bpnet_osknz_fold1   \n",
       "1         1       0  1.069548          85539188  oct4  bpnet_osknz_fold1   \n",
       "2         2       0  0.971074          85539189  oct4  bpnet_osknz_fold1   \n",
       "3         3       0  0.847520          85539190  oct4  bpnet_osknz_fold1   \n",
       "4         4       0  0.921221          85539191  oct4  bpnet_osknz_fold1   \n",
       "5         5       0  0.897421          85539192  oct4  bpnet_osknz_fold1   \n",
       "6         6       0  0.587497          85539193  oct4  bpnet_osknz_fold1   \n",
       "7         7       0  0.853134          85539194  oct4  bpnet_osknz_fold1   \n",
       "8         8       0  0.991755          85539195  oct4  bpnet_osknz_fold1   \n",
       "9         9       0  0.972739          85539196  oct4  bpnet_osknz_fold1   \n",
       "\n",
       "   scenario_index scenario state genomic_center_OS  genomic_center_S  \n",
       "0               0  WT_coop    AB          85539673          85539636  \n",
       "1               0  WT_coop    AB          85539673          85539636  \n",
       "2               0  WT_coop    AB          85539673          85539636  \n",
       "3               0  WT_coop    AB          85539673          85539636  \n",
       "4               0  WT_coop    AB          85539673          85539636  \n",
       "5               0  WT_coop    AB          85539673          85539636  \n",
       "6               0  WT_coop    AB          85539673          85539636  \n",
       "7               0  WT_coop    AB          85539673          85539636  \n",
       "8               0  WT_coop    AB          85539673          85539636  \n",
       "9               0  WT_coop    AB          85539673          85539636  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Convert logits and logcounts to human-readable ChIP-nexus profile with counts\n",
    "crispr_df = pd.DataFrame()\n",
    "models_of_interest = ['bpnet_osknz_fold1', 'atac_wt_fold1', 'atac_wt_fold2', 'atac_wt_fold3'] + \\\n",
    "[f'atac_{timepoint}h_fold1' for timepoint in concentration_atac_timepoints]\n",
    "for model_name in models_of_interest:\n",
    "    model_dir = modeling_design_dict[model_name]['model_dir'].replace('residual', 'combined')\n",
    "    acc_model = load_model(model_dir, custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})\n",
    "    crispr_preds = acc_model.predict(one_hot_encode_sequences(crispr_scenarios_df['inj_seq'].values))\n",
    "    \n",
    "    tasks = modeling_design_dict[model_name]['tasks']\n",
    "    tasks_n = len(tasks)\n",
    "    \n",
    "    for j,task in enumerate(tasks):\n",
    "        for i in crispr_scenarios_df['scenario_index'].values:\n",
    "            profile = logitsToProfile(logitsAcrossSingleRegion = crispr_preds[j][i], \n",
    "                                      logCountsAcrossSingleRegion = crispr_preds[j+tasks_n][i])\n",
    "            #Convert to tidy pd.df\n",
    "            df = pd.DataFrame(profile, columns = list(range(modeling_design_dict[model_name]['num-channels'])))\n",
    "            df['position'] = list(range(df.shape[0]))\n",
    "            df = df.melt(id_vars = 'position', var_name = 'channel', value_name = 'pred')\n",
    "            df['genomic_position'] = df['position'] + region_start + flank_length\n",
    "            df['task'] = task\n",
    "            df['model_name'] = model_name\n",
    "            df['scenario_index'] = i\n",
    "            crispr_df = pd.concat([crispr_df, df])\n",
    "crispr_df = crispr_df.merge(crispr_scenarios_df[['scenario_index', 'scenario', 'state', 'genomic_center_OS', 'genomic_center_S']], how = 'left')\n",
    "crispr_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_df.to_csv('tsv/genomic/crispr/crispr_coop_predictions_with_bias.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict effects of mutating out Zic3 motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "zic3_center = flank_length + 427\n",
    "zic3_mut = 'CATTGCAAAAG'\n",
    "crispr_scenarios_df['inj_seq_zic3_mut'] = [s[:(zic3_center-6)] + zic3_mut + s[(zic3_center+5):] for s in crispr_scenarios_df.inj_seq.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 346ms/step\n",
      "1/1 [==============================] - 0s 416ms/step\n",
      "1/1 [==============================] - 0s 410ms/step\n",
      "1/1 [==============================] - 0s 419ms/step\n",
      "1/1 [==============================] - 0s 425ms/step\n",
      "1/1 [==============================] - 0s 419ms/step\n",
      "1/1 [==============================] - 0s 430ms/step\n",
      "1/1 [==============================] - 0s 402ms/step\n",
      "1/1 [==============================] - 0s 411ms/step\n",
      "1/1 [==============================] - 0s 402ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>channel</th>\n",
       "      <th>pred</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>task</th>\n",
       "      <th>model_name</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>scenario</th>\n",
       "      <th>state</th>\n",
       "      <th>genomic_center_OS</th>\n",
       "      <th>genomic_center_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.422710</td>\n",
       "      <td>85539187</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.710536</td>\n",
       "      <td>85539188</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.624773</td>\n",
       "      <td>85539189</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.687584</td>\n",
       "      <td>85539190</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.735550</td>\n",
       "      <td>85539191</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.697468</td>\n",
       "      <td>85539192</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.484215</td>\n",
       "      <td>85539193</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.695770</td>\n",
       "      <td>85539194</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.797302</td>\n",
       "      <td>85539195</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.832765</td>\n",
       "      <td>85539196</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>WT_coop</td>\n",
       "      <td>AB</td>\n",
       "      <td>85539673</td>\n",
       "      <td>85539636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position channel      pred  genomic_position  task         model_name  \\\n",
       "0         0       0  0.422710          85539187  oct4  bpnet_osknz_fold1   \n",
       "1         1       0  0.710536          85539188  oct4  bpnet_osknz_fold1   \n",
       "2         2       0  0.624773          85539189  oct4  bpnet_osknz_fold1   \n",
       "3         3       0  0.687584          85539190  oct4  bpnet_osknz_fold1   \n",
       "4         4       0  0.735550          85539191  oct4  bpnet_osknz_fold1   \n",
       "5         5       0  0.697468          85539192  oct4  bpnet_osknz_fold1   \n",
       "6         6       0  0.484215          85539193  oct4  bpnet_osknz_fold1   \n",
       "7         7       0  0.695770          85539194  oct4  bpnet_osknz_fold1   \n",
       "8         8       0  0.797302          85539195  oct4  bpnet_osknz_fold1   \n",
       "9         9       0  0.832765          85539196  oct4  bpnet_osknz_fold1   \n",
       "\n",
       "   scenario_index scenario state genomic_center_OS  genomic_center_S  \n",
       "0               0  WT_coop    AB          85539673          85539636  \n",
       "1               0  WT_coop    AB          85539673          85539636  \n",
       "2               0  WT_coop    AB          85539673          85539636  \n",
       "3               0  WT_coop    AB          85539673          85539636  \n",
       "4               0  WT_coop    AB          85539673          85539636  \n",
       "5               0  WT_coop    AB          85539673          85539636  \n",
       "6               0  WT_coop    AB          85539673          85539636  \n",
       "7               0  WT_coop    AB          85539673          85539636  \n",
       "8               0  WT_coop    AB          85539673          85539636  \n",
       "9               0  WT_coop    AB          85539673          85539636  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Convert logits and logcounts to human-readable ChIP-nexus profile with counts\n",
    "crispr_df = pd.DataFrame()\n",
    "models_of_interest = ['bpnet_osknz_fold1', 'atac_wt_fold1', 'atac_wt_fold2', 'atac_wt_fold3'] + \\\n",
    "[f'atac_{timepoint}h_fold1' for timepoint in concentration_atac_timepoints]\n",
    "for model_name in models_of_interest:\n",
    "    model_dir = modeling_design_dict[model_name]['model_dir'].replace('residual', 'combined')\n",
    "    acc_model = load_model(model_dir, custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})\n",
    "    crispr_preds = acc_model.predict(one_hot_encode_sequences(crispr_scenarios_df['inj_seq_zic3_mut'].values))\n",
    "    \n",
    "    tasks = modeling_design_dict[model_name]['tasks']\n",
    "    tasks_n = len(tasks)\n",
    "    \n",
    "    for j,task in enumerate(tasks):\n",
    "        for i in crispr_scenarios_df['scenario_index'].values:\n",
    "            profile = logitsToProfile(logitsAcrossSingleRegion = crispr_preds[j][i], \n",
    "                                      logCountsAcrossSingleRegion = crispr_preds[j+tasks_n][i])\n",
    "            #Convert to tidy pd.df\n",
    "            df = pd.DataFrame(profile, columns = list(range(modeling_design_dict[model_name]['num-channels'])))\n",
    "            df['position'] = list(range(df.shape[0]))\n",
    "            df = df.melt(id_vars = 'position', var_name = 'channel', value_name = 'pred')\n",
    "            df['genomic_position'] = df['position'] + region_start + flank_length\n",
    "            df['task'] = task\n",
    "            df['model_name'] = model_name\n",
    "            df['scenario_index'] = i\n",
    "            crispr_df = pd.concat([crispr_df, df])\n",
    "crispr_df = crispr_df.merge(crispr_scenarios_df[['scenario_index', 'scenario', 'state', 'genomic_center_OS', 'genomic_center_S']], how = 'left')\n",
    "crispr_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_df.to_csv('tsv/genomic/crispr/crispr_coop_predictions_with_zic3_mut_with_bias.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate contribution scores for each mutation set\n",
    "\n",
    "First, obtain the sequences of interest and save as .fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = f'fasta/crispr_coop_seqs.fa'\n",
    "myfile = open(fasta_path, 'w')\n",
    "for i,row in crispr_scenarios_df.iterrows():\n",
    "    \n",
    "    header = '>' + row.scenario + row.state\n",
    "    myfile.write(\"%s\\n\" % header)\n",
    "    myfile.write(\"%s\\n\" % row.inj_seq)\n",
    "    \n",
    "myfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, define the counts output head by which to SHAP the scores back."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/crispr_coop_bpnet_shap.slurm\n"
     ]
    }
   ],
   "source": [
    "array_cmds = []\n",
    "\n",
    "for model_name,model_info in modeling_design_dict.items():\n",
    "\n",
    "    #Set up SHAP parameters for each task.\n",
    "    shap_template_dict = {\n",
    "        # 'genome': genome,\n",
    "        'fasta-file': fasta_path, \n",
    "        'input-length': input_length,\n",
    "        'output-length': output_length,\n",
    "        'num-shuffles': 20,\n",
    "        'verbosity': 'DEBUG'\n",
    "    }\n",
    "\n",
    "    #For ATAC vs CHIP-nexus, set up SHAP parameters\n",
    "    if model_info['tasks'][0]=='atac':\n",
    "        task = 'atac'\n",
    "        model_suffix = '_residual.model'\n",
    "        shap_dict = shap_template_dict.copy()\n",
    "        shap_dict['model-file'] = model_info['model_dir']\n",
    "        shap_dict['heads'] = 1\n",
    "        shap_dict['head-id'] = 0\n",
    "        shap_dict['profile-task-ids'] = [0]\n",
    "        shap_dict['profile-h5'] = f'shap/crispr/crispr_coop_{model_name}_{task}_profile.h5'\n",
    "        shap_dict['counts-h5'] = f'shap/crispr/crispr_coop_{model_name}_{task}_counts.h5'\n",
    "        shap_json = json.dumps(shap_dict, indent=4)\n",
    "        shap_json_file = f'json/crispr/shapFlat_crispr_coop_{model_name}_{task}.json'\n",
    "        with open(shap_json_file, 'w') as outfile:\n",
    "            outfile.write(shap_json)     \n",
    "\n",
    "        #Set up the SHAP commands\n",
    "        model_shap = [f'{python_path} {bpreveal_path}/src/interpretFlat.py json/crispr/shapFlat_crispr_coop_{model_name}_{task}.json']\n",
    "        # cmds = gpu_header + model_shap\n",
    "        # with open(f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm', mode='wt') as slurm:\n",
    "        #     slurm.write('\\n'.join(cmds))\n",
    "        #     slurm.write('\\n')\n",
    "        # print('sbatch ', f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm')\n",
    "        #Write to a job array\n",
    "        cmds_str = '\\n'.join(model_shap)\n",
    "        array_cmd = [cmds_str]\n",
    "        array_cmds += array_cmd\n",
    "        \n",
    "    else:\n",
    "        model_suffix = '.model'\n",
    "        shap_dict = shap_template_dict.copy()\n",
    "        shap_dict['model-file'] = f'models/{model_name}{model_suffix}'\n",
    "        for head_counter, (task, cov) in enumerate(model_info['cov'].items()):\n",
    "            shap_dict['heads'] = len(model_info['tasks'])\n",
    "            shap_dict['head-id'] = head_counter\n",
    "            shap_dict['profile-task-ids'] = [0,1]\n",
    "            shap_dict['profile-h5'] = f'shap/crispr/crispr_coop_{model_name}_{task}_profile.h5'\n",
    "            shap_dict['counts-h5'] = f'shap/crispr/crispr_coop_{model_name}_{task}_counts.h5'\n",
    "            shap_json = json.dumps(shap_dict, indent=4)\n",
    "            shap_json_file = f'json/crispr/shapFlat_crispr_coop_{model_name}_{task}.json'\n",
    "            with open(shap_json_file, 'w') as outfile:\n",
    "                outfile.write(shap_json)     \n",
    "    \n",
    "            #Set up the SHAP commands\n",
    "            model_shap = [f'{python_path} {bpreveal_path}/src/interpretFlat.py json/crispr/shapFlat_crispr_coop_{model_name}_{task}.json']\n",
    "            # cmds = gpu_header + model_shap\n",
    "            # with open(f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm', mode='wt') as slurm:\n",
    "            #     slurm.write('\\n'.join(cmds))\n",
    "            #     slurm.write('\\n')\n",
    "            # print('sbatch ', f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm')   \n",
    "\n",
    "            #Write to a job array\n",
    "            cmds_str = '\\n'.join(model_shap)\n",
    "            array_cmd = [cmds_str]\n",
    "            array_cmds += array_cmd\n",
    "\n",
    "generate_slurm_array(cmds = array_cmds, output_file = 'scripts/crispr_coop_bpnet_shap.slurm')\n",
    "print('sbatch scripts/crispr_coop_bpnet_shap.slurm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform PISA across each CRISPR set\n",
    "\n",
    "Define .json by which we want to perform PISA, We will perform PISA across each of these different configurations to ensure that OS/S motifs in this region are mostly cooperating with one another. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/bpnet_pisa_array_enhancers_across_crispr_coop.slurm\n"
     ]
    }
   ],
   "source": [
    "array_cmds = []\n",
    "for model_name,model_info in modeling_design_dict.items():\n",
    "    tasks = model_info['tasks']\n",
    "\n",
    "    enhancer_pisa_template_dict = {\n",
    "        \"genome\": genome,\n",
    "        'bed-file': 'tsv/genomic/crispr/crispr_coop_coord.bed', #'fasta-file': fasta_path, \n",
    "        \"model-file\": model_info['model_dir'],\n",
    "        \"input-length\": input_length,\n",
    "        \"output-length\": output_length,\n",
    "        \"heads\": len(tasks),\n",
    "        \"num-shuffles\": 20,\n",
    "        \"verbosity\": \"DEBUG\"\n",
    "    }\n",
    "\n",
    "    for head_counter, task in enumerate(tasks):\n",
    "        for channel_counter in range(model_info['num-channels']):\n",
    "            enhancer_pisa_dict = enhancer_pisa_template_dict.copy()\n",
    "            enhancer_pisa_dict['head-id'] = head_counter\n",
    "            enhancer_pisa_dict['task-id'] = channel_counter\n",
    "            enhancer_pisa_dict['output-h5'] = f\"pisa/crispr_coop_pisa_{model_name}_{task}_{channel_counter}.h5\"\n",
    "            enhancer_pisa_json = json.dumps(enhancer_pisa_dict, indent=4)\n",
    "            enhancer_pisa_file = f'json/interpretPISA_crispr_coop_{model_name}_{task}_{channel_counter}.json'\n",
    "    \n",
    "            with open(enhancer_pisa_file, 'w') as outfile:\n",
    "                outfile.write(enhancer_pisa_json)\n",
    "                \n",
    "            #Set up training and prediction commands\n",
    "            pisa = [f'{python_path} {bpreveal_path}/src/interpretPisa.py json/interpretPISA_crispr_coop_{model_name}_{task}_{channel_counter}.json']\n",
    "    \n",
    "            #Write to a job array\n",
    "            cmds_str = '\\n'.join(pisa)\n",
    "            array_cmd = [cmds_str]\n",
    "            array_cmds += array_cmd\n",
    "\n",
    "generate_slurm_array(cmds = array_cmds, output_file = 'scripts/bpnet_pisa_array_enhancers_across_crispr_coop.slurm')\n",
    "print('sbatch scripts/bpnet_pisa_array_enhancers_across_crispr_coop.slurm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perturb Akr1cl site with 2 identical Sox2 motifs\n",
    "\n",
    "This site was identified by (1) having 2 Sox2 motifs with identical sequences within the same nucleosome distance (2) these two motifs needed to show opposing functions of pioneering vs non-pioneering. This showcases that motif affinity is not the obly driver of pioneering, and that context also matters. \n",
    "\n",
    "In this case, motifA will be the distal Sox2 and motifB will be the proximal Sox2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     motif seqnames     wt_seq  motif_window_start  motif_window_end  \\\n",
      "4290  Sox2     chr1  CCCTTTGTC                 864               873   \n",
      "\n",
      "      pattern_center  \n",
      "4290             868  \n",
      "     motif seqnames     wt_seq  motif_window_start  motif_window_end  \\\n",
      "4292  Sox2     chr1  CCCTTTGTC                 963               972   \n",
      "\n",
      "      pattern_center  \n",
      "4292             967  \n"
     ]
    }
   ],
   "source": [
    "sox2_distal_motif_id = 253082\n",
    "sox2_proximal_motif_id = 253084\n",
    "region_id = 3491\n",
    "\n",
    "#Isolate motifs\n",
    "motifs_df = pd.read_csv(motifs_path,  sep = '\\t')\n",
    "motifs_df['pattern_center'] = motifs_df['pattern_center'] + flank_length\n",
    "sox2_distal_motif_df = motifs_df[motifs_df.motif_id==sox2_distal_motif_id]\n",
    "sox2_distal_motif_df['wt_seq'] = extract_seqs_from_df(sox2_distal_motif_df, genome)\n",
    "sox2_proximal_motif_df = motifs_df[motifs_df.motif_id==sox2_proximal_motif_id]\n",
    "sox2_proximal_motif_df['wt_seq'] = extract_seqs_from_df(sox2_proximal_motif_df, genome)\n",
    "\n",
    "#Keep metadata of use\n",
    "columns_of_interest = ['motif', 'seqnames', 'wt_seq', 'motif_window_start', 'motif_window_end', 'pattern_center']\n",
    "sox2_distal_motif_df = sox2_distal_motif_df[columns_of_interest]\n",
    "sox2_proximal_motif_df = sox2_proximal_motif_df[columns_of_interest]\n",
    "print(sox2_distal_motif_df)\n",
    "print(sox2_proximal_motif_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Print region start coordinate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65036790"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_start = motifs_df[motifs_df.motif_id==sox2_distal_motif_id].region_start.values[0]\n",
    "region_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overwrite distance injection for `dist_coop` scenario."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current center-to-center distance:  [99]\n"
     ]
    }
   ],
   "source": [
    "print('Current center-to-center distance: ', sox2_proximal_motif_df.pattern_center.values - sox2_distal_motif_df.pattern_center.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define distal Sox2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define all CRISPR scenarios for the distal Sox2. Sox2 will not change in position or be enhanced by affinity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq</th>\n",
       "      <th>motif</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>wt_seq</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "      <th>pattern_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  state forward_seq motif seqnames     wt_seq  motif_window_start  \\\n",
       "0    AB   CCCTTTGTC  Sox2     chr1  CCCTTTGTC                 864   \n",
       "1     B   CCCTAGGTC  Sox2     chr1  CCCTTTGTC                 864   \n",
       "2     A   CCCTTTGTC  Sox2     chr1  CCCTTTGTC                 864   \n",
       "\n",
       "   motif_window_end  pattern_center  \n",
       "0               873             868  \n",
       "1               873             868  \n",
       "2               873             868  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sox2_distal_states_df = pd.DataFrame([['AB', 'B', 'A'], ['CCCTTTGTC', 'CCCTAGGTC', 'CCCTTTGTC']])\\\n",
    "    .transpose()\\\n",
    "    .rename({0: 'state', 1:'forward_seq'}, axis = 1)\n",
    "sox2_distal_states_df['motif'] = 'Sox2'\n",
    "sox2_distal_states_df = sox2_distal_states_df.merge(sox2_distal_motif_df)\n",
    "sox2_distal_states_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define proximal Sox2\n",
    "\n",
    "Define all CRISPR scenarios for the proximal Sox2. Sox2 will not change in position or be enhanced by affinity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq</th>\n",
       "      <th>motif</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>wt_seq</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "      <th>pattern_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  state forward_seq motif seqnames     wt_seq  motif_window_start  \\\n",
       "0    AB   CCCTTTGTC  Sox2     chr1  CCCTTTGTC                 963   \n",
       "1     A   CCCTAGGTC  Sox2     chr1  CCCTTTGTC                 963   \n",
       "2     B   CCCTTTGTC  Sox2     chr1  CCCTTTGTC                 963   \n",
       "\n",
       "   motif_window_end  pattern_center  \n",
       "0               972             967  \n",
       "1               972             967  \n",
       "2               972             967  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sox2_proximal_states_df = pd.DataFrame([['AB', 'A', 'B'], ['CCCTTTGTC', 'CCCTAGGTC', 'CCCTTTGTC']])\\\n",
    "    .transpose()\\\n",
    "    .rename({0: 'state', 1:'forward_seq'}, axis = 1)\n",
    "sox2_proximal_states_df['motif'] = 'Sox2'\n",
    "sox2_proximal_states_df = sox2_proximal_states_df.merge(sox2_proximal_motif_df)\n",
    "sox2_proximal_states_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge coordinates together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq_distal</th>\n",
       "      <th>motif_distal</th>\n",
       "      <th>seqnames_distal</th>\n",
       "      <th>wt_seq_distal</th>\n",
       "      <th>motif_window_start_distal</th>\n",
       "      <th>motif_window_end_distal</th>\n",
       "      <th>pattern_center_distal</th>\n",
       "      <th>forward_seq_proximal</th>\n",
       "      <th>motif_proximal</th>\n",
       "      <th>seqnames_proximal</th>\n",
       "      <th>wt_seq_proximal</th>\n",
       "      <th>motif_window_start_proximal</th>\n",
       "      <th>motif_window_end_proximal</th>\n",
       "      <th>pattern_center_proximal</th>\n",
       "      <th>genomic_start_distal</th>\n",
       "      <th>genomic_end_distal</th>\n",
       "      <th>genomic_center_distal</th>\n",
       "      <th>genomic_start_proximal</th>\n",
       "      <th>genomic_end_proximal</th>\n",
       "      <th>genomic_center_proximal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  state forward_seq_distal motif_distal seqnames_distal wt_seq_distal  \\\n",
       "0    AB          CCCTTTGTC         Sox2            chr1     CCCTTTGTC   \n",
       "1     B          CCCTAGGTC         Sox2            chr1     CCCTTTGTC   \n",
       "2     A          CCCTTTGTC         Sox2            chr1     CCCTTTGTC   \n",
       "\n",
       "   motif_window_start_distal  motif_window_end_distal  pattern_center_distal  \\\n",
       "0                        864                      873                    868   \n",
       "1                        864                      873                    868   \n",
       "2                        864                      873                    868   \n",
       "\n",
       "  forward_seq_proximal motif_proximal seqnames_proximal wt_seq_proximal  \\\n",
       "0            CCCTTTGTC           Sox2              chr1       CCCTTTGTC   \n",
       "1            CCCTTTGTC           Sox2              chr1       CCCTTTGTC   \n",
       "2            CCCTAGGTC           Sox2              chr1       CCCTTTGTC   \n",
       "\n",
       "   motif_window_start_proximal  motif_window_end_proximal  \\\n",
       "0                          963                        972   \n",
       "1                          963                        972   \n",
       "2                          963                        972   \n",
       "\n",
       "   pattern_center_proximal  genomic_start_distal  genomic_end_distal  \\\n",
       "0                      967              65037654            65037663   \n",
       "1                      967              65037654            65037663   \n",
       "2                      967              65037654            65037663   \n",
       "\n",
       "   genomic_center_distal  genomic_start_proximal  genomic_end_proximal  \\\n",
       "0               65037658                65037753              65037762   \n",
       "1               65037658                65037753              65037762   \n",
       "2               65037658                65037753              65037762   \n",
       "\n",
       "   genomic_center_proximal  \n",
       "0                 65037757  \n",
       "1                 65037757  \n",
       "2                 65037757  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_scenarios_df = sox2_distal_states_df.merge(sox2_proximal_states_df, how = 'left', on = 'state', suffixes=('_distal', '_proximal'))\n",
    "crispr_scenarios_df[['genomic_start_distal', 'genomic_end_distal', 'genomic_center_distal']] = crispr_scenarios_df[['motif_window_start_distal', 'motif_window_end_distal', 'pattern_center_distal']] + region_start\n",
    "crispr_scenarios_df[['genomic_start_proximal', 'genomic_end_proximal', 'genomic_center_proximal']] = crispr_scenarios_df[['motif_window_start_proximal', 'motif_window_end_proximal', 'pattern_center_proximal']] + region_start\n",
    "crispr_scenarios_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate predictions based on the different scenarios\n",
    "\n",
    "First, prepare WT genomic sequences according to input accessibility features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_region_df = regions_df[regions_df.region_id==region_id]\n",
    "crispr_region_df['wt_seq'] = extract_seqs_from_df(crispr_region_df, genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chrom     start       end  region_id  score strand  \\\n",
      "3491  chr1  65036790  65038822       3491      0      .   \n",
      "\n",
      "                                                 wt_seq  \n",
      "3491  TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...  \n",
      "TTGATGTTAG\n",
      "AGCATAAAAG\n"
     ]
    }
   ],
   "source": [
    "print(crispr_region_df)\n",
    "print(crispr_region_df.wt_seq.values[0][:10])\n",
    "print(crispr_region_df.wt_seq.values[0][-10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_region_df.to_csv('tsv/genomic/crispr/crispr_context_coord.bed', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_model = load_model(modeling_design_dict['atac_wt_fold1']['model_dir'], custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, inject motifs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_seq = crispr_region_df.wt_seq.values[0]\n",
    "\n",
    "crispr_seqs = []\n",
    "for i,row in crispr_scenarios_df.iterrows():\n",
    "        #Overwrite proximal Sox2\n",
    "        s_seq = wt_seq[:(row.motif_window_start_proximal)] + row.forward_seq_proximal + wt_seq[(row.motif_window_end_proximal):]\n",
    "        \n",
    "        #Overwrite distal Sox2\n",
    "        final_seq = s_seq[:(row.motif_window_start_distal)] + row.forward_seq_distal + s_seq[(row.motif_window_end_distal):]\n",
    "        assert len(final_seq)==input_length\n",
    "        crispr_seqs.append(final_seq)\n",
    "\n",
    "crispr_scenarios_df['inj_seq'] = crispr_seqs\n",
    "crispr_scenarios_df['scenario_index'] = list(range(crispr_scenarios_df.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>state</th>\n",
       "      <th>forward_seq_distal</th>\n",
       "      <th>motif_distal</th>\n",
       "      <th>seqnames_distal</th>\n",
       "      <th>wt_seq_distal</th>\n",
       "      <th>motif_window_start_distal</th>\n",
       "      <th>motif_window_end_distal</th>\n",
       "      <th>pattern_center_distal</th>\n",
       "      <th>forward_seq_proximal</th>\n",
       "      <th>motif_proximal</th>\n",
       "      <th>seqnames_proximal</th>\n",
       "      <th>wt_seq_proximal</th>\n",
       "      <th>motif_window_start_proximal</th>\n",
       "      <th>motif_window_end_proximal</th>\n",
       "      <th>pattern_center_proximal</th>\n",
       "      <th>genomic_start_distal</th>\n",
       "      <th>genomic_end_distal</th>\n",
       "      <th>genomic_center_distal</th>\n",
       "      <th>genomic_start_proximal</th>\n",
       "      <th>genomic_end_proximal</th>\n",
       "      <th>genomic_center_proximal</th>\n",
       "      <th>inj_seq</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>inj_seq_65037600_to_65037900</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "      <td>TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "      <td>TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>864</td>\n",
       "      <td>873</td>\n",
       "      <td>868</td>\n",
       "      <td>CCCTAGGTC</td>\n",
       "      <td>Sox2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CCCTTTGTC</td>\n",
       "      <td>963</td>\n",
       "      <td>972</td>\n",
       "      <td>967</td>\n",
       "      <td>65037654</td>\n",
       "      <td>65037663</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037753</td>\n",
       "      <td>65037762</td>\n",
       "      <td>65037757</td>\n",
       "      <td>TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...</td>\n",
       "      <td>2</td>\n",
       "      <td>AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  state forward_seq_distal motif_distal seqnames_distal wt_seq_distal  \\\n",
       "0    AB          CCCTTTGTC         Sox2            chr1     CCCTTTGTC   \n",
       "1     B          CCCTAGGTC         Sox2            chr1     CCCTTTGTC   \n",
       "2     A          CCCTTTGTC         Sox2            chr1     CCCTTTGTC   \n",
       "\n",
       "   motif_window_start_distal  motif_window_end_distal  pattern_center_distal  \\\n",
       "0                        864                      873                    868   \n",
       "1                        864                      873                    868   \n",
       "2                        864                      873                    868   \n",
       "\n",
       "  forward_seq_proximal motif_proximal seqnames_proximal wt_seq_proximal  \\\n",
       "0            CCCTTTGTC           Sox2              chr1       CCCTTTGTC   \n",
       "1            CCCTTTGTC           Sox2              chr1       CCCTTTGTC   \n",
       "2            CCCTAGGTC           Sox2              chr1       CCCTTTGTC   \n",
       "\n",
       "   motif_window_start_proximal  motif_window_end_proximal  \\\n",
       "0                          963                        972   \n",
       "1                          963                        972   \n",
       "2                          963                        972   \n",
       "\n",
       "   pattern_center_proximal  genomic_start_distal  genomic_end_distal  \\\n",
       "0                      967              65037654            65037663   \n",
       "1                      967              65037654            65037663   \n",
       "2                      967              65037654            65037663   \n",
       "\n",
       "   genomic_center_distal  genomic_start_proximal  genomic_end_proximal  \\\n",
       "0               65037658                65037753              65037762   \n",
       "1               65037658                65037753              65037762   \n",
       "2               65037658                65037753              65037762   \n",
       "\n",
       "   genomic_center_proximal                                            inj_seq  \\\n",
       "0                 65037757  TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...   \n",
       "1                 65037757  TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...   \n",
       "2                 65037757  TTGATGTTAGCCATTTGATTGGTGTAAGGCAGAATTTTTAGCTTTG...   \n",
       "\n",
       "   scenario_index                       inj_seq_65037600_to_65037900  \n",
       "0               0  AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...  \n",
       "1               1  AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...  \n",
       "2               2  AACTGCTGCTATTTTCTGACGGGCCACATCTGTCTCCTTTCTGTCC...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_scenarios_df['inj_seq_65037600_to_65037900'] = [s[(65037600 - crispr_region_df.start.values[0]):(65037600 - crispr_region_df.start.values[0] + 300) ]\n",
    "                                                       for s in crispr_scenarios_df['inj_seq'].values]\n",
    "crispr_scenarios_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_scenarios_df.to_csv('tsv/genomic/crispr/crispr_context_scenarios.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CCCTTTGTC', 'CCCTAGGTC', 'CCCTTTGTC']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_distal_sox2_start = crispr_scenarios_df.motif_window_start_distal.values[0]\n",
    "original_distal_sox2_end = crispr_scenarios_df.motif_window_end_distal.values[0]\n",
    "[s[(original_distal_sox2_start):(original_distal_sox2_end)] for s in crispr_scenarios_df.inj_seq.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CCCTTTGTC', 'CCCTTTGTC', 'CCCTAGGTC']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_proximal_sox2_start = crispr_scenarios_df.motif_window_start_proximal.values[0]\n",
    "original_proximal_sox2_end = crispr_scenarios_df.motif_window_end_proximal.values[0]\n",
    "[s[(original_proximal_sox2_start):(original_proximal_sox2_end)] for s in crispr_scenarios_df.inj_seq.values]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict injected sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 275ms/step\n",
      "1/1 [==============================] - 0s 278ms/step\n",
      "1/1 [==============================] - 0s 272ms/step\n",
      "1/1 [==============================] - 0s 217ms/step\n",
      "1/1 [==============================] - 0s 220ms/step\n",
      "1/1 [==============================] - 0s 229ms/step\n",
      "1/1 [==============================] - 0s 225ms/step\n",
      "1/1 [==============================] - 0s 216ms/step\n",
      "1/1 [==============================] - 0s 221ms/step\n",
      "1/1 [==============================] - 0s 217ms/step\n",
      "1/1 [==============================] - 0s 205ms/step\n",
      "1/1 [==============================] - 0s 215ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>channel</th>\n",
       "      <th>pred</th>\n",
       "      <th>task</th>\n",
       "      <th>model_name</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>state</th>\n",
       "      <th>genomic_center_distal</th>\n",
       "      <th>genomic_center_proximal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>65037306</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005036</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>65037307</td>\n",
       "      <td>0</td>\n",
       "      <td>0.009420</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>65037308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033758</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>65037309</td>\n",
       "      <td>0</td>\n",
       "      <td>0.029952</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>65037310</td>\n",
       "      <td>0</td>\n",
       "      <td>0.015480</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>65037311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005962</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>65037312</td>\n",
       "      <td>0</td>\n",
       "      <td>0.006315</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>65037313</td>\n",
       "      <td>0</td>\n",
       "      <td>0.004534</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>65037314</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008796</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>65037315</td>\n",
       "      <td>0</td>\n",
       "      <td>0.013199</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position  genomic_position channel      pred  task         model_name  \\\n",
       "0         0          65037306       0  0.005036  oct4  bpnet_osknz_fold1   \n",
       "1         1          65037307       0  0.009420  oct4  bpnet_osknz_fold1   \n",
       "2         2          65037308       0  0.033758  oct4  bpnet_osknz_fold1   \n",
       "3         3          65037309       0  0.029952  oct4  bpnet_osknz_fold1   \n",
       "4         4          65037310       0  0.015480  oct4  bpnet_osknz_fold1   \n",
       "5         5          65037311       0  0.005962  oct4  bpnet_osknz_fold1   \n",
       "6         6          65037312       0  0.006315  oct4  bpnet_osknz_fold1   \n",
       "7         7          65037313       0  0.004534  oct4  bpnet_osknz_fold1   \n",
       "8         8          65037314       0  0.008796  oct4  bpnet_osknz_fold1   \n",
       "9         9          65037315       0  0.013199  oct4  bpnet_osknz_fold1   \n",
       "\n",
       "   scenario_index state  genomic_center_distal  genomic_center_proximal  \n",
       "0               0    AB               65037658                 65037757  \n",
       "1               0    AB               65037658                 65037757  \n",
       "2               0    AB               65037658                 65037757  \n",
       "3               0    AB               65037658                 65037757  \n",
       "4               0    AB               65037658                 65037757  \n",
       "5               0    AB               65037658                 65037757  \n",
       "6               0    AB               65037658                 65037757  \n",
       "7               0    AB               65037658                 65037757  \n",
       "8               0    AB               65037658                 65037757  \n",
       "9               0    AB               65037658                 65037757  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Convert logits and logcounts to human-readable ChIP-nexus profile with counts\n",
    "crispr_df = pd.DataFrame()\n",
    "for model_name in list(modeling_design_dict.keys()):\n",
    "    model = load_model(modeling_design_dict[model_name]['model_dir'], \n",
    "                           custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})\n",
    "    crispr_preds = model.predict(one_hot_encode_sequences(crispr_scenarios_df['inj_seq'].values))\n",
    "    \n",
    "    for i in crispr_scenarios_df['scenario_index'].values:\n",
    "        tasks = modeling_design_dict[model_name]['tasks']\n",
    "        for j,task in enumerate(tasks):\n",
    "            profile = logitsToProfile(logitsAcrossSingleRegion = crispr_preds[j][i], \n",
    "                                      logCountsAcrossSingleRegion = crispr_preds[j + len(tasks)][i])\n",
    "            #Convert to tidy pd.df\n",
    "            df = pd.DataFrame(profile, columns = list(range(modeling_design_dict[model_name]['num-channels'])))\n",
    "            df['position'] = list(range(df.shape[0]))\n",
    "            df['genomic_position'] = df['position'] + region_start + flank_length\n",
    "            df = df.melt(id_vars = ['position', 'genomic_position'], var_name = 'channel', value_name = 'pred')\n",
    "            df['task'] = task\n",
    "            df['model_name'] = model_name\n",
    "            df['scenario_index'] = i\n",
    "            crispr_df = pd.concat([crispr_df, df])\n",
    "crispr_df = crispr_df.merge(crispr_scenarios_df[['scenario_index', 'state', 'genomic_center_distal', 'genomic_center_proximal']], how = 'left')\n",
    "crispr_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_df.to_csv('tsv/genomic/crispr/crispr_context_predictions.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict biased sequences for comparison with the experimental predictions in a fair manner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 279ms/step\n",
      "1/1 [==============================] - 0s 260ms/step\n",
      "1/1 [==============================] - 0s 276ms/step\n",
      "1/1 [==============================] - 0s 332ms/step\n",
      "1/1 [==============================] - 0s 327ms/step\n",
      "1/1 [==============================] - 0s 338ms/step\n",
      "1/1 [==============================] - 0s 347ms/step\n",
      "1/1 [==============================] - 0s 335ms/step\n",
      "1/1 [==============================] - 0s 344ms/step\n",
      "1/1 [==============================] - 0s 337ms/step\n",
      "1/1 [==============================] - 0s 351ms/step\n",
      "1/1 [==============================] - 0s 337ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>channel</th>\n",
       "      <th>pred</th>\n",
       "      <th>task</th>\n",
       "      <th>model_name</th>\n",
       "      <th>scenario_index</th>\n",
       "      <th>state</th>\n",
       "      <th>genomic_center_distal</th>\n",
       "      <th>genomic_center_proximal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>65037306</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005036</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>65037307</td>\n",
       "      <td>0</td>\n",
       "      <td>0.009420</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>65037308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.033758</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>65037309</td>\n",
       "      <td>0</td>\n",
       "      <td>0.029952</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>65037310</td>\n",
       "      <td>0</td>\n",
       "      <td>0.015480</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>65037311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005962</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>65037312</td>\n",
       "      <td>0</td>\n",
       "      <td>0.006315</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>65037313</td>\n",
       "      <td>0</td>\n",
       "      <td>0.004534</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>65037314</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008796</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>65037315</td>\n",
       "      <td>0</td>\n",
       "      <td>0.013199</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz_fold1</td>\n",
       "      <td>0</td>\n",
       "      <td>AB</td>\n",
       "      <td>65037658</td>\n",
       "      <td>65037757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position  genomic_position channel      pred  task         model_name  \\\n",
       "0         0          65037306       0  0.005036  oct4  bpnet_osknz_fold1   \n",
       "1         1          65037307       0  0.009420  oct4  bpnet_osknz_fold1   \n",
       "2         2          65037308       0  0.033758  oct4  bpnet_osknz_fold1   \n",
       "3         3          65037309       0  0.029952  oct4  bpnet_osknz_fold1   \n",
       "4         4          65037310       0  0.015480  oct4  bpnet_osknz_fold1   \n",
       "5         5          65037311       0  0.005962  oct4  bpnet_osknz_fold1   \n",
       "6         6          65037312       0  0.006315  oct4  bpnet_osknz_fold1   \n",
       "7         7          65037313       0  0.004534  oct4  bpnet_osknz_fold1   \n",
       "8         8          65037314       0  0.008796  oct4  bpnet_osknz_fold1   \n",
       "9         9          65037315       0  0.013199  oct4  bpnet_osknz_fold1   \n",
       "\n",
       "   scenario_index state  genomic_center_distal  genomic_center_proximal  \n",
       "0               0    AB               65037658                 65037757  \n",
       "1               0    AB               65037658                 65037757  \n",
       "2               0    AB               65037658                 65037757  \n",
       "3               0    AB               65037658                 65037757  \n",
       "4               0    AB               65037658                 65037757  \n",
       "5               0    AB               65037658                 65037757  \n",
       "6               0    AB               65037658                 65037757  \n",
       "7               0    AB               65037658                 65037757  \n",
       "8               0    AB               65037658                 65037757  \n",
       "9               0    AB               65037658                 65037757  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Convert logits and logcounts to human-readable ChIP-nexus profile with counts\n",
    "crispr_df = pd.DataFrame()\n",
    "for model_name in list(modeling_design_dict.keys()):\n",
    "    model_dir = modeling_design_dict[model_name]['model_dir'].replace('residual', 'combined')\n",
    "    model = load_model(model_dir, custom_objects = {'multinomialNll' : losses.multinomialNll, 'reweightableMse': losses.dummyMse})\n",
    "    crispr_preds = model.predict(one_hot_encode_sequences(crispr_scenarios_df['inj_seq'].values))\n",
    "    \n",
    "    for i in crispr_scenarios_df['scenario_index'].values:\n",
    "        tasks = modeling_design_dict[model_name]['tasks']\n",
    "        for j,task in enumerate(tasks):\n",
    "            profile = logitsToProfile(logitsAcrossSingleRegion = crispr_preds[j][i], \n",
    "                                      logCountsAcrossSingleRegion = crispr_preds[j + len(tasks)][i])\n",
    "            #Convert to tidy pd.df\n",
    "            df = pd.DataFrame(profile, columns = list(range(modeling_design_dict[model_name]['num-channels'])))\n",
    "            df['position'] = list(range(df.shape[0]))\n",
    "            df['genomic_position'] = df['position'] + region_start + flank_length\n",
    "            df = df.melt(id_vars = ['position', 'genomic_position'], var_name = 'channel', value_name = 'pred')\n",
    "            df['task'] = task\n",
    "            df['model_name'] = model_name\n",
    "            df['scenario_index'] = i\n",
    "            crispr_df = pd.concat([crispr_df, df])\n",
    "crispr_df = crispr_df.merge(crispr_scenarios_df[['scenario_index', 'state', 'genomic_center_distal', 'genomic_center_proximal']], how = 'left')\n",
    "crispr_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_df.to_csv('tsv/genomic/crispr/crispr_context_predictions_with_bias.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate contribution scores for each mutation set\n",
    "\n",
    "First, obtain the sequences of interest and save as .fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = f'fasta/crispr_context_seqs.fa'\n",
    "myfile = open(fasta_path, 'w')\n",
    "for i,row in crispr_scenarios_df.iterrows():\n",
    "    \n",
    "    header = '>' + row.state\n",
    "    myfile.write(\"%s\\n\" % header)\n",
    "    myfile.write(\"%s\\n\" % row.inj_seq)\n",
    "    \n",
    "myfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, define the counts output head by which to SHAP the scores back."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/crispr_context_bpnet_shap.slurm\n"
     ]
    }
   ],
   "source": [
    "array_cmds = []\n",
    "\n",
    "for model_name,model_info in modeling_design_dict.items():\n",
    "\n",
    "    #Set up SHAP parameters for each task.\n",
    "    shap_template_dict = {\n",
    "        # 'genome': genome,\n",
    "        'fasta-file': fasta_path, \n",
    "        'input-length': input_length,\n",
    "        'output-length': output_length,\n",
    "        'num-shuffles': 20,\n",
    "        'verbosity': 'DEBUG'\n",
    "    }\n",
    "\n",
    "    #For ATAC vs CHIP-nexus, set up SHAP parameters\n",
    "    if model_info['tasks'][0]=='atac':\n",
    "        task = 'atac'\n",
    "        model_suffix = '_residual.model'\n",
    "        shap_dict = shap_template_dict.copy()\n",
    "        shap_dict['model-file'] = model_info['model_dir']\n",
    "        shap_dict['heads'] = 1\n",
    "        shap_dict['head-id'] = 0\n",
    "        shap_dict['profile-task-ids'] = [0]\n",
    "        shap_dict['profile-h5'] = f'shap/crispr/crispr_context_{model_name}_{task}_profile.h5'\n",
    "        shap_dict['counts-h5'] = f'shap/crispr/crispr_context_{model_name}_{task}_counts.h5'\n",
    "        shap_json = json.dumps(shap_dict, indent=4)\n",
    "        shap_json_file = f'json/crispr/shapFlat_crispr_context_{model_name}_{task}.json'\n",
    "        with open(shap_json_file, 'w') as outfile:\n",
    "            outfile.write(shap_json)     \n",
    "\n",
    "        #Set up the SHAP commands\n",
    "        model_shap = [f'{python_path} {bpreveal_path}/src/interpretFlat.py json/crispr/shapFlat_crispr_context_{model_name}_{task}.json']\n",
    "        # cmds = gpu_header + model_shap\n",
    "        # with open(f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm', mode='wt') as slurm:\n",
    "        #     slurm.write('\\n'.join(cmds))\n",
    "        #     slurm.write('\\n')\n",
    "        # print('sbatch ', f'scripts/crispr_coop_bpnet_shap_{model_name}_{task}.slurm')\n",
    "        #Write to a job array\n",
    "        cmds_str = '\\n'.join(model_shap)\n",
    "        array_cmd = [cmds_str]\n",
    "        array_cmds += array_cmd\n",
    "        \n",
    "    else:\n",
    "        model_suffix = '.model'\n",
    "        shap_dict = shap_template_dict.copy()\n",
    "        shap_dict['model-file'] = f'models/{model_name}{model_suffix}'\n",
    "        for head_counter, (task, cov) in enumerate(model_info['cov'].items()):\n",
    "            shap_dict['heads'] = len(model_info['tasks'])\n",
    "            shap_dict['head-id'] = head_counter\n",
    "            shap_dict['profile-task-ids'] = [0,1]\n",
    "            shap_dict['profile-h5'] = f'shap/crispr/crispr_context_{model_name}_{task}_profile.h5'\n",
    "            shap_dict['counts-h5'] = f'shap/crispr/crispr_context_{model_name}_{task}_counts.h5'\n",
    "            shap_json = json.dumps(shap_dict, indent=4)\n",
    "            shap_json_file = f'json/crispr/shapFlat_crispr_context_{model_name}_{task}.json'\n",
    "            with open(shap_json_file, 'w') as outfile:\n",
    "                outfile.write(shap_json)     \n",
    "    \n",
    "            #Set up the SHAP commands\n",
    "            model_shap = [f'{python_path} {bpreveal_path}/src/interpretFlat.py json/crispr/shapFlat_crispr_context_{model_name}_{task}.json']\n",
    "\n",
    "            #Write to a job array\n",
    "            cmds_str = '\\n'.join(model_shap)\n",
    "            array_cmd = [cmds_str]\n",
    "            array_cmds += array_cmd\n",
    "\n",
    "generate_slurm_array(cmds = array_cmds, output_file = 'scripts/crispr_context_bpnet_shap.slurm')\n",
    "print('sbatch scripts/crispr_context_bpnet_shap.slurm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bpreveal_404] *",
   "language": "python",
   "name": "conda-env-bpreveal_404-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
