{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The goals of this analysis is to perform in genomic perturbations and analysis on the ATAC and OSKNZ binding models. Here we seek to:\n",
    "\n",
    "+ assess the impact of all mapped motifs on binding and accessibility\n",
    "+ assess the impact of all mapped motif pairs on binding and accessibility\n",
    "+ assess the profile changes of mapped motif perturbations across selected enhancers\n",
    "\n",
    "# Computational setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-07-02 09:13:38.873473: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2025-07-02 09:13:38.873513: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2025-07-02 09:13:38.874919: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2025-07-02 09:13:38.882132: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-07-02 09:13:40.022237: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import keras\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import keras.backend as K\n",
    "from keras.models import load_model\n",
    "from pybedtools import BedTool\n",
    "from tqdm import tqdm\n",
    "import pickle as pkl\n",
    "from glob import glob\n",
    "\n",
    "# Settings\n",
    "\n",
    "## Working options\n",
    "os.chdir(f'/n/projects/mw2098/publications/2024_weilert_acc/code/2_analysis/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "bpreveal_path = '/n/projects/mw2098/publications/2024_weilert_acc/public/software/bpreveal_404/'\n",
    "python_path = '/home/mw2098/anaconda3/envs/bpreveal_404/bin/python'\n",
    "\n",
    "sys.path.insert(0, f'{bpreveal_path}/src')\n",
    "import losses\n",
    "\n",
    "## Custom functions\n",
    "sys.path.insert(0, f'scripts/py/functions')\n",
    "from functional import shuffle_seqs, one_hot_encode_sequence, one_hot_encode_sequences, \\\n",
    "    one_hot_decode_sequence, insert_motif, logitsToProfile\n",
    "from motifs import extract_seqs_from_df, resize_coordinates\n",
    "from perturb import generate_alt_sequences\n",
    "\n",
    "## Custom variables\n",
    "figure_path = 'figures/9_perturb_motifs'\n",
    "working_dir = os.getcwd()\n",
    "genome = '../0_setup/fa/mm10.fa'\n",
    "motifs_path = 'bed/mapped_motifs/all_instances_curated_0based.bed'\n",
    "xiong_motifs_path = 'bed/mapped_motifs/all_xiong_instances_curated_0based.bed'\n",
    "\n",
    "motifs_w_metadata_path = 'tsv/mapped_motifs/all_instances_curated_1based.tsv.gz'\n",
    "xiong_motifs_w_metadata_path = 'tsv/mapped_motifs/all_xiong_instances_curated_1based.tsv.gz'\n",
    "regions_path = 'bed/mapped_motifs/all_islands_curated_0based_sized_to_input.bed'\n",
    "xiong_regions_path = 'bed/mapped_motifs/all_xiong_islands_curated_0based_sized_to_input.bed'\n",
    "showcase_medians_path = 'tsv/insilico/marginalizations/motifs/all_median_profiles.tsv.gz'\n",
    "\n",
    "motif_to_task_dict = {'Oct4-Sox2': 'oct4', \n",
    "                      'Oct4': 'oct4',\n",
    "                      'Sox2': 'sox2',\n",
    "                      'Klf4': 'klf4',\n",
    "                      'Zic3': 'zic3',\n",
    "                      'Nanog': 'nanog'}\n",
    "region_dict = {'oct4': 'narrowpeak/mesc_oct4_nexus_peaks.narrowPeak',\n",
    "               'sox2': 'narrowpeak/mesc_sox2_nexus_peaks.narrowPeak',\n",
    "               'klf4': 'narrowpeak/mesc_klf4_nexus_peaks.narrowPeak',\n",
    "               'nanog': 'narrowpeak/mesc_nanog_nexus_peaks.narrowPeak',\n",
    "               'zic3': 'narrowpeak/mesc_zic3_nexus_peaks.narrowPeak'}\n",
    "concentration_atac_timepoints = list(range(0, 16, 3))\n",
    "modeling_design_dict = {\n",
    "    'bpnet_osknz':{\n",
    "        'tasks': list(region_dict.keys()),\n",
    "        'cov': {k: {'pos': 'bw/mesc_' + k + '_nexus_combined_positive.bw',\n",
    "                    'neg': 'bw/mesc_' + k + '_nexus_combined_negative.bw'}\n",
    "                for k in region_dict.keys()},\n",
    "        'model_dir': 'models/bpnet_osknz_fold1.model',\n",
    "        'num-channels': 2\n",
    "    },\n",
    "    'atac_wt':{\n",
    "        'tasks': ['atac'],\n",
    "        'cov': 'bw/mesc_native_atac_cutsites_combined.bw',\n",
    "        'model_dir': 'models/atac_wt_fold1_residual.model/',\n",
    "        'num-channels': 1\n",
    "    }\n",
    "}\n",
    "for timepoint in concentration_atac_timepoints:\n",
    "    modeling_design_dict[f'atac_{timepoint}h'] = {\n",
    "            'tasks': ['atac'],\n",
    "            'cov': f'bw/GSE174774_mesc_atac_{timepoint}h_combined.bw',\n",
    "            'model_dir': f'models/atac_{timepoint}h_fold1_residual.model/',\n",
    "            'num-channels': 1\n",
    "    }\n",
    "input_length = 2032\n",
    "output_length = 1000\n",
    "trials = 16\n",
    "\n",
    "## Filesystem commands\n",
    "!mkdir -p {figure_path}\n",
    "!mkdir -p tsv/genomic/motifs\n",
    "!mkdir -p scripts/genomic/motifs\n",
    "!mkdir -p tsv/genomic/enhancers\n",
    "!mkdir -p pisa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Custom functions\n",
    "\n",
    "### Function to queue slurm jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_slurm_array(cmds, output_file, simultaneous_jobs = 4):\n",
    "    total_jobs = len(cmds)\n",
    "    array_header = ['#!/usr/bin/bash',\n",
    "                  '#SBATCH --job-name bpnet',\n",
    "                  '#SBATCH --output=slurm_%j.log',\n",
    "                  f'#SBATCH --array=1-{total_jobs}%{simultaneous_jobs}',\n",
    "                  '#SBATCH --mem=200gb', \n",
    "                  '#SBATCH --time=72:00:00',\n",
    "                  '#SBATCH --partition=gpu', \n",
    "                  '#SBATCH --gres=gpu:a100:1',\n",
    "                  '#SBATCH --cpus-per-task=10',\n",
    "                  'source /home/mw2098/.bashrc',\n",
    "                  'conda deactivate',\n",
    "                  'conda activate bpreveal_404',\n",
    "                  f'cd {working_dir}']\n",
    "    # array_cmds = array_header\n",
    "    for i,cmd in enumerate(cmds):\n",
    "        array_cmd = [\"if [[ ${{SLURM_ARRAY_TASK_ID}} == {0:d} ]] ; then\\n\".format(i+1), \n",
    "                    \"    {0:s}\\n\".format(cmd),\n",
    "                    \"fi\\n\\n\"]\n",
    "        array_header = array_header + array_cmd \n",
    "        \n",
    "    with open(output_file, mode='wt') as slurm:\n",
    "        slurm.write('\\n'.join(array_header))\n",
    "        slurm.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to generate script for genomic perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_genomic_perturb_script(motifs_df_path, motif_groups_df_path, \n",
    "                                 model_dir, fasta_file,\n",
    "                                 tasks_in_model_order,\n",
    "                                 output_prefix, output_cmd_path, \n",
    "                                 region_index_column = 'region_id', \n",
    "                                 motif_window_start_column = 'motif_window_start',\n",
    "                                 motif_window_end_column = 'motif_window_end',\n",
    "                                 motif_name_column = 'motif',\n",
    "                                 motif_chrom_column = 'chrom',\n",
    "                                 summary_window = 'use_whole_window', #`use_whole_window`, `keep_entire_profile`, or int\n",
    "                                 comb_max = 2, comb_min = 1, \n",
    "                                 nodes = 24, trials = 16, \n",
    "                                 input_seqlen = 2114, \n",
    "                                 output_seqlen = 1000):\n",
    "    \n",
    "    working_dir = os.getcwd()\n",
    "    gpu_header = ['#!/bin/bash',\n",
    "                      '#SBATCH --job-name genomic_perturbations',\n",
    "                      '#SBATCH --output=slurm_%j.log',\n",
    "                      '#SBATCH --mem=500gb', '#SBATCH --time=72:00:00',\n",
    "                      '#SBATCH --partition=gpu', '#SBATCH --gres=gpu:a100:1',\n",
    "                      '#SBATCH --cpus-per-task=30',\n",
    "                      'source /home/mw2098/.bashrc',\n",
    "                      'conda deactivate',\n",
    "                      'conda activate bpreveal_404',\n",
    "                      f'cd {working_dir}']\n",
    "    cmd = [\n",
    "        f'{python_path} scripts/py/tools/generate_genomic_perturbs.py \\\\',\n",
    "        f'--motifs_df_path {motifs_df_path} \\\\',\n",
    "        f'--motif_groups_df_path {motif_groups_df_path} \\\\',\n",
    "        f'--model_dir {model_dir} \\\\',\n",
    "        f'--output_prefix {output_prefix} \\\\',\n",
    "        f'--tasks_in_model_order {tasks_in_model_order} \\\\',\n",
    "        f'--region_index_column {region_index_column} \\\\',\n",
    "        f'--motif_window_start_column {motif_window_start_column} \\\\',\n",
    "        f'--motif_window_end_column {motif_window_end_column} \\\\',\n",
    "        f'--motif_name_column {motif_name_column} \\\\',\n",
    "        f'--motif_chrom_column {motif_chrom_column} \\\\',\n",
    "        f'--fasta_file {fasta_file} \\\\',\n",
    "        f'--input_seqlen {input_seqlen} --output_seqlen {output_seqlen} \\\\',\n",
    "        f'--comb_max {comb_max} --comb_min {comb_min} \\\\',\n",
    "        f'--nodes {nodes} --trials {trials} \\\\'\n",
    "    ]\n",
    "    if type(summary_window)==int:\n",
    "        last_param = [f'--summary_window {summary_window} ']\n",
    "        cmd = cmd + (list(last_param))      \n",
    "    else:\n",
    "        last_param = [f'--{summary_window} ']\n",
    "        cmd = cmd + (list(last_param))\n",
    "    cmds = gpu_header + cmd\n",
    "    \n",
    "    #Write script\n",
    "    output_hit_mapping_script = open(output_cmd_path, \"w\")\n",
    "    for i in cmds:\n",
    "        output_hit_mapping_script.write(i + \"\\n\")\n",
    "    output_hit_mapping_script.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform genomic perturbations and summarize effects\n",
    "\n",
    "## Format motifs for perturbations\n",
    "\n",
    "To do this, we must first consolidate and organize motifs.\n",
    "\n",
    "Here, we need to import the motifs that were curated and add the correct columns such that the bpnet_generate_perturbations script can be satisfied. To do this, we need a 0-based coordinate .tsv file with the following columns: `motif`, `region_id`, `chrom`, `motif_window_start`, `motif_window_end`, `motif_length`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(377687, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_length</th>\n",
       "      <th>motif</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>region_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154541</td>\n",
       "      <td>3154556</td>\n",
       "      <td>Oct4-Sox2_118382_4</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118382</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155627</td>\n",
       "      <td>3155642</td>\n",
       "      <td>Oct4-Sox2_118383_5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118383</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155712</td>\n",
       "      <td>3155727</td>\n",
       "      <td>Oct4-Sox2_118384_5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118384</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3263877</td>\n",
       "      <td>3263892</td>\n",
       "      <td>Oct4-Sox2_118385_7</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118385</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3343669</td>\n",
       "      <td>3343684</td>\n",
       "      <td>Oct4-Sox2_118386_9</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118386</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end                name  score strand  motif_length  \\\n",
       "0  chr1  3154541  3154556  Oct4-Sox2_118382_4      0      +            15   \n",
       "1  chr1  3155627  3155642  Oct4-Sox2_118383_5      0      +            15   \n",
       "2  chr1  3155712  3155727  Oct4-Sox2_118384_5      0      +            15   \n",
       "3  chr1  3263877  3263892  Oct4-Sox2_118385_7      0      -            15   \n",
       "4  chr1  3343669  3343684  Oct4-Sox2_118386_9      0      +            15   \n",
       "\n",
       "       motif motif_id region_id  \n",
       "0  Oct4-Sox2   118382         4  \n",
       "1  Oct4-Sox2   118383         5  \n",
       "2  Oct4-Sox2   118384         5  \n",
       "3  Oct4-Sox2   118385         7  \n",
       "4  Oct4-Sox2   118386         9  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import set of motifs\n",
    "motifs_df = BedTool(motifs_path).to_dataframe()\n",
    "motifs_df.columns = ['chrom','start','end','name','score','strand']\n",
    "\n",
    "#Separate motif name\n",
    "motifs_df['motif_length'] = motifs_df['end'] - motifs_df['start']\n",
    "motifs_df['motif'] = [n.split('_')[0] for n in motifs_df.name]\n",
    "motifs_df['motif_id'] = [n.split('_')[1] for n in motifs_df.name]\n",
    "motifs_df['region_id'] = [n.split('_')[2] for n in motifs_df.name]\n",
    "print(motifs_df.shape)\n",
    "\n",
    "motifs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format regions in order to mark the internal motif window coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_df = BedTool(regions_path).to_dataframe()\n",
    "regions_df.columns = ['region_chrom','region_start','region_end','region_id','region_score','region_strand']\n",
    "regions_df['region_id']=regions_df['region_id'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Match grouped regions to motifs and collect motif position within the windows to obtain `motif_window_start` and `motif_window_end`. We do this now because it is far easier to work with 0-based coordinated in python than in R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs_df = motifs_df.merge(regions_df[['region_start','region_end','region_id']], on = 'region_id', how = 'left')\n",
    "motifs_df['motif_window_start'] = motifs_df['start']-motifs_df['region_start']\n",
    "motifs_df['motif_window_end'] = motifs_df['end']-motifs_df['region_start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(377687, 14)\n"
     ]
    }
   ],
   "source": [
    "motifs_df.head(n=10)\n",
    "print(motifs_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter motifs on edges of the windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(377687, 14)\n"
     ]
    }
   ],
   "source": [
    "motifs_df = motifs_df[((motifs_df['start']  - motifs_df['region_start'])>=20) & ((motifs_df['region_end']  - motifs_df['end'])>=20)]\n",
    "print(motifs_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter motifs that are of poor `fann` similarity matching to seqlet cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(377687, 42)\n"
     ]
    }
   ],
   "source": [
    "motifs_w_metadata_df = pd.read_csv(motifs_w_metadata_path, sep = '\\t').drop(columns = ['start','end','motif', 'strand'])\n",
    "motifs_df['motif_id'] = motifs_df.motif_id.astype(int)\n",
    "motifs_df = motifs_df.merge(motifs_w_metadata_df, how = 'left', on = 'motif_id', suffixes=('', '_y'))\n",
    "print(motifs_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif\n",
      "Klf4         118382\n",
      "Zic3          92751\n",
      "Nanog         65959\n",
      "Sox2          59781\n",
      "Oct4-Sox2     40682\n",
      "Oct4            132\n",
      "Name: count, dtype: int64\n",
      "count\n",
      "1     80995\n",
      "2     33923\n",
      "3     20888\n",
      "4     13138\n",
      "5      8353\n",
      "6      4835\n",
      "7      2647\n",
      "8      1387\n",
      "9       680\n",
      "10      329\n",
      "11      152\n",
      "12       64\n",
      "13       40\n",
      "14       19\n",
      "15        6\n",
      "18        4\n",
      "22        3\n",
      "17        3\n",
      "21        2\n",
      "16        2\n",
      "20        1\n",
      "27        1\n",
      "25        1\n",
      "24        1\n",
      "46        1\n",
      "36        1\n",
      "35        1\n",
      "28        1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Save motifs\n",
    "motifs_df.to_csv(f'tsv/genomic/motifs/instances_formatted_for_genomic_perturbations_0based.tsv.gz', sep = '\\t', index = False)\n",
    "print(motifs_df.motif.value_counts())\n",
    "print(motifs_df.region_id.value_counts().value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  region_chrom  region_start  region_end region_id  region_score  \\\n",
      "0         chr1       3034951     3036983         0             0   \n",
      "1         chr1       3051532     3053564         1             0   \n",
      "2         chr1       3058391     3060423         2             0   \n",
      "3         chr1       3061941     3063973         3             0   \n",
      "4         chr1       3153574     3155606         4             0   \n",
      "\n",
      "  region_strand  island_count  \n",
      "0             .             3  \n",
      "1             .             3  \n",
      "2             .             5  \n",
      "3             .             4  \n",
      "4             .             3  \n"
     ]
    }
   ],
   "source": [
    "#Save regions and rename regions_df columns to match more standard bed formatting for accessibility perturbations\n",
    "regions_df = regions_df.merge(motifs_df[['region_id','island_count']].drop_duplicates(ignore_index = True, keep = 'first'), \n",
    "                              on = 'region_id', how = 'left')\n",
    "print(regions_df.head())\n",
    "regions_df.columns = ['chrom', 'start', 'end', 'region_id', 'score', 'strand', 'island_count']\n",
    "regions_df.to_csv(f'tsv/genomic/motifs/regions_formatted_for_genomic_perturbations_0based.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate binding and accessibility perturbations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Allow combinations of up to 2 and up to 3 mutations. This is promarily used for cooperativity and comparisons among island effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_bpnet_osknz.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_wt.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_0h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_3h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_6h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_9h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_12h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_atac_15h.slurm\n"
     ]
    }
   ],
   "source": [
    "for pertubation_name, perturbation_details in modeling_design_dict.items():\n",
    "    write_genomic_perturb_script(motifs_df_path = f'tsv/genomic/motifs/instances_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 motif_groups_df_path = f'tsv/genomic/motifs/regions_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 model_dir = perturbation_details['model_dir'], \n",
    "                                 fasta_file = genome,\n",
    "                                 tasks_in_model_order = ','.join(perturbation_details['tasks']),\n",
    "                                 output_prefix = f'tsv/genomic/motifs/genomic_pertubs_whole_window_{pertubation_name}', \n",
    "                                 output_cmd_path = f'scripts/genomic/motifs/genomic_pertubs_whole_window_{pertubation_name}.slurm', \n",
    "                                 input_seqlen = input_length, \n",
    "                                 output_seqlen = output_length,\n",
    "                                 trials = trials,\n",
    "                                 comb_max = 2, comb_min = 1)\n",
    "    print(f'sbatch scripts/genomic/motifs/genomic_pertubs_whole_window_{pertubation_name}.slurm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write another genomic perturbation script with windowed limitations to calculate directional cooperativity for the binding model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/genomic/motifs/genomic_pertubs_75bp_window_bpnet_osknz.slurm\n"
     ]
    }
   ],
   "source": [
    "write_genomic_perturb_script(motifs_df_path = f'tsv/genomic/motifs/instances_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 motif_groups_df_path = f'tsv/genomic/motifs/regions_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 model_dir = modeling_design_dict['bpnet_osknz']['model_dir'], \n",
    "                                 fasta_file = genome,\n",
    "                                 tasks_in_model_order = ','.join(modeling_design_dict['bpnet_osknz']['tasks']),\n",
    "                                 output_prefix = f'tsv/genomic/motifs/genomic_pertubs_75bp_window_bpnet_osknz', \n",
    "                                 output_cmd_path = f'scripts/genomic/motifs/genomic_pertubs_75bp_window_bpnet_osknz.slurm', \n",
    "                                 input_seqlen = input_length, \n",
    "                                 output_seqlen = output_length,\n",
    "                                 trials = trials,\n",
    "                                 summary_window = 75,\n",
    "                                 comb_max = 2, comb_min = 1)\n",
    "print(f'sbatch scripts/genomic/motifs/genomic_pertubs_75bp_window_bpnet_osknz.slurm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_length</th>\n",
       "      <th>motif</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>region_id</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>width</th>\n",
       "      <th>short_name</th>\n",
       "      <th>metacluster_name</th>\n",
       "      <th>pattern_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>region_index</th>\n",
       "      <th>seq_match</th>\n",
       "      <th>seq_match_quantile</th>\n",
       "      <th>task</th>\n",
       "      <th>model</th>\n",
       "      <th>mapping_state</th>\n",
       "      <th>seq</th>\n",
       "      <th>region_id_y</th>\n",
       "      <th>region_start_1based</th>\n",
       "      <th>region_end_1based</th>\n",
       "      <th>island_content</th>\n",
       "      <th>island_content_ordered</th>\n",
       "      <th>island_content_unique</th>\n",
       "      <th>island_count</th>\n",
       "      <th>acc_contrib</th>\n",
       "      <th>oct4_contrib</th>\n",
       "      <th>sox2_contrib</th>\n",
       "      <th>nanog_contrib</th>\n",
       "      <th>klf4_contrib</th>\n",
       "      <th>zic3_contrib</th>\n",
       "      <th>bind_contrib</th>\n",
       "      <th>is_across_erv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154541</td>\n",
       "      <td>3154556</td>\n",
       "      <td>Oct4-Sox2_118382_4</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118382</td>\n",
       "      <td>4</td>\n",
       "      <td>3153574</td>\n",
       "      <td>3155606</td>\n",
       "      <td>967</td>\n",
       "      <td>982</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15</td>\n",
       "      <td>pos_0</td>\n",
       "      <td>pos_patterns</td>\n",
       "      <td>pattern_0</td>\n",
       "      <td>ATTTGCATTTGAATA</td>\n",
       "      <td>11</td>\n",
       "      <td>13.426259</td>\n",
       "      <td>0.972220</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz</td>\n",
       "      <td>both</td>\n",
       "      <td>ATTTGCATTTGAATA</td>\n",
       "      <td>4</td>\n",
       "      <td>3154091</td>\n",
       "      <td>3155090</td>\n",
       "      <td>Klf4:1_Oct4-Sox2:1_Zic3:1</td>\n",
       "      <td>Zic3_Oct4-Sox2_Klf4</td>\n",
       "      <td>Klf4_Oct4-Sox2_Zic3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.234222</td>\n",
       "      <td>0.416462</td>\n",
       "      <td>0.137217</td>\n",
       "      <td>0.131604</td>\n",
       "      <td>0.027325</td>\n",
       "      <td>0.026796</td>\n",
       "      <td>0.416462</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155627</td>\n",
       "      <td>3155642</td>\n",
       "      <td>Oct4-Sox2_118383_5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118383</td>\n",
       "      <td>5</td>\n",
       "      <td>3154660</td>\n",
       "      <td>3156692</td>\n",
       "      <td>967</td>\n",
       "      <td>982</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15</td>\n",
       "      <td>pos_0</td>\n",
       "      <td>pos_patterns</td>\n",
       "      <td>pattern_0</td>\n",
       "      <td>ATTAGCATTATAAAG</td>\n",
       "      <td>12</td>\n",
       "      <td>12.903305</td>\n",
       "      <td>0.915456</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz</td>\n",
       "      <td>both</td>\n",
       "      <td>ATTAGCATTATAAAG</td>\n",
       "      <td>5</td>\n",
       "      <td>3155177</td>\n",
       "      <td>3156176</td>\n",
       "      <td>Oct4-Sox2:2</td>\n",
       "      <td>Oct4-Sox2_Oct4-Sox2</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.283484</td>\n",
       "      <td>0.120707</td>\n",
       "      <td>0.043296</td>\n",
       "      <td>0.185654</td>\n",
       "      <td>-0.007187</td>\n",
       "      <td>0.044399</td>\n",
       "      <td>0.120707</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155712</td>\n",
       "      <td>3155727</td>\n",
       "      <td>Oct4-Sox2_118384_5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118384</td>\n",
       "      <td>5</td>\n",
       "      <td>3154660</td>\n",
       "      <td>3156692</td>\n",
       "      <td>1052</td>\n",
       "      <td>1067</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15</td>\n",
       "      <td>pos_0</td>\n",
       "      <td>pos_patterns</td>\n",
       "      <td>pattern_0</td>\n",
       "      <td>TTTACCATAAGAAAA</td>\n",
       "      <td>12</td>\n",
       "      <td>11.593916</td>\n",
       "      <td>0.540766</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz</td>\n",
       "      <td>bind</td>\n",
       "      <td>TTTACCATAAGAAAA</td>\n",
       "      <td>5</td>\n",
       "      <td>3155177</td>\n",
       "      <td>3156176</td>\n",
       "      <td>Oct4-Sox2:2</td>\n",
       "      <td>Oct4-Sox2_Oct4-Sox2</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.127052</td>\n",
       "      <td>0.255989</td>\n",
       "      <td>0.102353</td>\n",
       "      <td>0.179189</td>\n",
       "      <td>0.026077</td>\n",
       "      <td>0.048198</td>\n",
       "      <td>0.255989</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3263877</td>\n",
       "      <td>3263892</td>\n",
       "      <td>Oct4-Sox2_118385_7</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118385</td>\n",
       "      <td>7</td>\n",
       "      <td>3262904</td>\n",
       "      <td>3264936</td>\n",
       "      <td>973</td>\n",
       "      <td>988</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15</td>\n",
       "      <td>pos_0</td>\n",
       "      <td>pos_patterns</td>\n",
       "      <td>pattern_0</td>\n",
       "      <td>AGTGGCATATCAAAG</td>\n",
       "      <td>13</td>\n",
       "      <td>11.694463</td>\n",
       "      <td>0.587819</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz</td>\n",
       "      <td>bind</td>\n",
       "      <td>AGTGGCATATCAAAG</td>\n",
       "      <td>7</td>\n",
       "      <td>3263421</td>\n",
       "      <td>3264420</td>\n",
       "      <td>Nanog:1_Oct4-Sox2:1_Zic3:1</td>\n",
       "      <td>Zic3_Oct4-Sox2_Nanog</td>\n",
       "      <td>Nanog_Oct4-Sox2_Zic3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.165599</td>\n",
       "      <td>0.196452</td>\n",
       "      <td>0.129591</td>\n",
       "      <td>0.259290</td>\n",
       "      <td>0.117235</td>\n",
       "      <td>0.244606</td>\n",
       "      <td>0.196452</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3343669</td>\n",
       "      <td>3343684</td>\n",
       "      <td>Oct4-Sox2_118386_9</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>15</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>118386</td>\n",
       "      <td>9</td>\n",
       "      <td>3342695</td>\n",
       "      <td>3344727</td>\n",
       "      <td>974</td>\n",
       "      <td>989</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15</td>\n",
       "      <td>pos_0</td>\n",
       "      <td>pos_patterns</td>\n",
       "      <td>pattern_0</td>\n",
       "      <td>ATTTACAACTGAATA</td>\n",
       "      <td>15</td>\n",
       "      <td>11.007610</td>\n",
       "      <td>0.354589</td>\n",
       "      <td>oct4</td>\n",
       "      <td>bpnet_osknz</td>\n",
       "      <td>both</td>\n",
       "      <td>ATTTACAACTGAATA</td>\n",
       "      <td>9</td>\n",
       "      <td>3343212</td>\n",
       "      <td>3344211</td>\n",
       "      <td>Nanog:2_Oct4-Sox2:1</td>\n",
       "      <td>Nanog_Oct4-Sox2_Nanog</td>\n",
       "      <td>Nanog_Oct4-Sox2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.726059</td>\n",
       "      <td>0.346188</td>\n",
       "      <td>0.214333</td>\n",
       "      <td>0.460609</td>\n",
       "      <td>0.156565</td>\n",
       "      <td>0.198544</td>\n",
       "      <td>0.346188</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end                name  score strand  motif_length  \\\n",
       "0  chr1  3154541  3154556  Oct4-Sox2_118382_4      0      +            15   \n",
       "1  chr1  3155627  3155642  Oct4-Sox2_118383_5      0      +            15   \n",
       "2  chr1  3155712  3155727  Oct4-Sox2_118384_5      0      +            15   \n",
       "3  chr1  3263877  3263892  Oct4-Sox2_118385_7      0      -            15   \n",
       "4  chr1  3343669  3343684  Oct4-Sox2_118386_9      0      +            15   \n",
       "\n",
       "       motif  motif_id region_id  region_start  region_end  \\\n",
       "0  Oct4-Sox2    118382         4       3153574     3155606   \n",
       "1  Oct4-Sox2    118383         5       3154660     3156692   \n",
       "2  Oct4-Sox2    118384         5       3154660     3156692   \n",
       "3  Oct4-Sox2    118385         7       3262904     3264936   \n",
       "4  Oct4-Sox2    118386         9       3342695     3344727   \n",
       "\n",
       "   motif_window_start  motif_window_end seqnames  width short_name  \\\n",
       "0                 967               982     chr1     15      pos_0   \n",
       "1                 967               982     chr1     15      pos_0   \n",
       "2                1052              1067     chr1     15      pos_0   \n",
       "3                 973               988     chr1     15      pos_0   \n",
       "4                 974               989     chr1     15      pos_0   \n",
       "\n",
       "  metacluster_name pattern_name         sequence  region_index  seq_match  \\\n",
       "0     pos_patterns    pattern_0  ATTTGCATTTGAATA            11  13.426259   \n",
       "1     pos_patterns    pattern_0  ATTAGCATTATAAAG            12  12.903305   \n",
       "2     pos_patterns    pattern_0  TTTACCATAAGAAAA            12  11.593916   \n",
       "3     pos_patterns    pattern_0  AGTGGCATATCAAAG            13  11.694463   \n",
       "4     pos_patterns    pattern_0  ATTTACAACTGAATA            15  11.007610   \n",
       "\n",
       "   seq_match_quantile  task        model mapping_state              seq  \\\n",
       "0            0.972220  oct4  bpnet_osknz          both  ATTTGCATTTGAATA   \n",
       "1            0.915456  oct4  bpnet_osknz          both  ATTAGCATTATAAAG   \n",
       "2            0.540766  oct4  bpnet_osknz          bind  TTTACCATAAGAAAA   \n",
       "3            0.587819  oct4  bpnet_osknz          bind  AGTGGCATATCAAAG   \n",
       "4            0.354589  oct4  bpnet_osknz          both  ATTTACAACTGAATA   \n",
       "\n",
       "   region_id_y  region_start_1based  region_end_1based  \\\n",
       "0            4              3154091            3155090   \n",
       "1            5              3155177            3156176   \n",
       "2            5              3155177            3156176   \n",
       "3            7              3263421            3264420   \n",
       "4            9              3343212            3344211   \n",
       "\n",
       "               island_content island_content_ordered island_content_unique  \\\n",
       "0   Klf4:1_Oct4-Sox2:1_Zic3:1    Zic3_Oct4-Sox2_Klf4   Klf4_Oct4-Sox2_Zic3   \n",
       "1                 Oct4-Sox2:2    Oct4-Sox2_Oct4-Sox2             Oct4-Sox2   \n",
       "2                 Oct4-Sox2:2    Oct4-Sox2_Oct4-Sox2             Oct4-Sox2   \n",
       "3  Nanog:1_Oct4-Sox2:1_Zic3:1   Zic3_Oct4-Sox2_Nanog  Nanog_Oct4-Sox2_Zic3   \n",
       "4         Nanog:2_Oct4-Sox2:1  Nanog_Oct4-Sox2_Nanog       Nanog_Oct4-Sox2   \n",
       "\n",
       "   island_count  acc_contrib  oct4_contrib  sox2_contrib  nanog_contrib  \\\n",
       "0             3     0.234222      0.416462      0.137217       0.131604   \n",
       "1             2     0.283484      0.120707      0.043296       0.185654   \n",
       "2             2     0.127052      0.255989      0.102353       0.179189   \n",
       "3             3     0.165599      0.196452      0.129591       0.259290   \n",
       "4             3     0.726059      0.346188      0.214333       0.460609   \n",
       "\n",
       "   klf4_contrib  zic3_contrib  bind_contrib  is_across_erv  \n",
       "0      0.027325      0.026796      0.416462           True  \n",
       "1     -0.007187      0.044399      0.120707           True  \n",
       "2      0.026077      0.048198      0.255989           True  \n",
       "3      0.117235      0.244606      0.196452           True  \n",
       "4      0.156565      0.198544      0.346188           True  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predict effects when all motifs are mutated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_bpnet_osknz.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_wt.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_0h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_3h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_6h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_9h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_12h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_atac_15h.slurm\n"
     ]
    }
   ],
   "source": [
    "for pertubation_name, perturbation_details in modeling_design_dict.items():\n",
    "    write_genomic_perturb_script(motifs_df_path = f'tsv/genomic/motifs/instances_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 motif_groups_df_path = f'tsv/genomic/motifs/regions_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 model_dir = perturbation_details['model_dir'], \n",
    "                                 fasta_file = genome,\n",
    "                                 tasks_in_model_order = ','.join(perturbation_details['tasks']),\n",
    "                                 output_prefix = f'tsv/genomic/motifs/genomic_pertubs_all_motifs_{pertubation_name}', \n",
    "                                 output_cmd_path = f'scripts/genomic/motifs/genomic_pertubs_all_motifs_{pertubation_name}.slurm', \n",
    "                                 input_seqlen = input_length, \n",
    "                                 output_seqlen = output_length,\n",
    "                                 trials = trials,\n",
    "                                 comb_max = 'island_count', comb_min = 'island_count')\n",
    "    print(f'sbatch scripts/genomic/motifs/genomic_pertubs_all_motifs_{pertubation_name}.slurm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter median motifs for showcase isolation/context contrast\n",
    "\n",
    "In previous scripts, we filtered the center 100 motifs based on median isolations score signals to depict the profiles across different motifs. We will take all motifs that fit those exact sequences and keep the profiles of contribution scores. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "median_seqs_df = pd.read_csv(showcase_medians_path, sep = '\\t')\n",
    "median_motifs_df = motifs_df[motifs_df['seq'].isin(median_seqs_df.seq.values)].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif\n",
      "Nanog        5853\n",
      "Klf4          549\n",
      "Sox2          363\n",
      "Zic3          358\n",
      "Oct4           27\n",
      "Oct4-Sox2      26\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Save motifs\n",
    "median_motifs_df.to_csv(f'tsv/genomic/motifs/medians_formatted_for_genomic_perturbations_0based.tsv.gz', sep = '\\t', index = False)\n",
    "print(median_motifs_df.motif.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate binding and accessibility profiles for median motif sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_bpnet_osknz.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_wt.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_0h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_3h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_6h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_9h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_12h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_atac_15h.slurm\n"
     ]
    }
   ],
   "source": [
    "for pertubation_name, perturbation_details in modeling_design_dict.items():\n",
    "    write_genomic_perturb_script(motifs_df_path = f'tsv/genomic/motifs/medians_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 motif_groups_df_path = f'tsv/genomic/motifs/regions_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 model_dir = perturbation_details['model_dir'], \n",
    "                                 fasta_file = genome,\n",
    "                                 tasks_in_model_order = ','.join(perturbation_details['tasks']),\n",
    "                                 output_prefix = f'tsv/genomic/motifs/genomic_pertubs_median_motifs_{pertubation_name}', \n",
    "                                 output_cmd_path = f'scripts/genomic/motifs/genomic_pertubs_median_motifs_{pertubation_name}.slurm', \n",
    "                                 input_seqlen = input_length, \n",
    "                                 output_seqlen = output_length,\n",
    "                                 trials = trials,\n",
    "                                 summary_window = 'keep_entire_profile',\n",
    "                                 comb_max = 1, comb_min = 1)\n",
    "    print(f'sbatch scripts/genomic/motifs/genomic_pertubs_median_motifs_{pertubation_name}.slurm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert .pkl file to .tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████| 20/20 [13:35<00:00, 40.76s/it]\n",
      "100%|████████████████████████████████████████████████████████| 20/20 [00:09<00:00,  2.11it/s]\n",
      "100%|████████████████████████████████████████████████████████| 20/20 [00:09<00:00,  2.06it/s]\n",
      "100%|████████████████████████████████████████████████████████| 20/20 [00:10<00:00,  1.96it/s]\n",
      "100%|████████████████████████████████████████████████████████| 20/20 [00:09<00:00,  2.03it/s]\n",
      "100%|████████████████████████████████████████████████████████| 20/20 [00:10<00:00,  1.96it/s]\n",
      "100%|████████████████████████| 20/20 [00:09<00:00,  2.04it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:09<00:00,  2.13it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13999000, 8)\n"
     ]
    }
   ],
   "source": [
    "for pertubation_name, perturbation_details in modeling_design_dict.items():\n",
    "    tasks = perturbation_details['tasks']\n",
    "\n",
    "    preds_all_df = pd.DataFrame()\n",
    "    index_paths = glob(f'tsv/genomic/motifs/genomic_pertubs_median_motifs_{pertubation_name}_index_*.tsv.gz')\n",
    "    \n",
    "    for index_path in tqdm(index_paths):\n",
    "        pred_path = index_path.replace('_index', '').replace('.tsv.gz', '.pkl')\n",
    "\n",
    "        indexes_df = pd.read_csv(index_path, sep = '\\t')\n",
    "        with open(pred_path, 'rb') as f:\n",
    "            preds_dict = pkl.load(f)\n",
    "\n",
    "        if type(perturbation_details['cov']) == str:\n",
    "            preds_arr = np.squeeze(preds_dict['atac'])\n",
    "            preds_df = pd.concat([indexes_df, pd.DataFrame(preds_arr, columns=range(preds_arr.shape[1]))], axis = 1) \\\n",
    "                .melt(id_vars = indexes_df.columns, var_name = 'window_position', value_name = 'pred')\n",
    "            preds_df['model_name'] = pertubation_name            \n",
    "            preds_df['task'] = 'atac'            \n",
    "            preds_df['channel'] = 'atac'            \n",
    "            preds_all_df = pd.concat([preds_all_df, preds_df])\n",
    "        else:\n",
    "            for task in tasks:\n",
    "                for i,channel in enumerate(list(perturbation_details['cov'][task].keys())):\n",
    "                    preds_arr = np.squeeze(preds_dict[task][:,:,i])\n",
    "                    preds_df = pd.concat([indexes_df, pd.DataFrame(preds_arr, columns=range(preds_arr.shape[1]))], axis = 1) \\\n",
    "                        .melt(id_vars = indexes_df.columns, var_name = 'window_position', value_name = 'pred')\n",
    "                    preds_df['model_name'] = pertubation_name            \n",
    "                    preds_df['task'] = task            \n",
    "                    preds_df['channel'] = channel            \n",
    "                    preds_all_df = pd.concat([preds_all_df, preds_df])\n",
    "                \n",
    "    preds_all_df.to_csv(f'tsv/genomic/motifs/genomic_pertubs_median_motifs_formatted_{pertubation_name}.tsv.gz', sep = '\\t', index = False)\n",
    "\n",
    "print(preds_all_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize this alongside isolation scores of median values in later .Rmd files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Xiong genomic perturbations and summarize effects\n",
    "\n",
    "For the motifs mapped by the 0h timepoint from the ATAC-seq time course from Xiong et al, perform perturbations on those explicit motifs.\n",
    "\n",
    "## Format motifs for perturbations\n",
    "\n",
    "To do this, we must first consolidate and organize motifs.\n",
    "\n",
    "Here, we need to import the motifs that were curated and add the correct columns such that the bpnet_generate_perturbations script can be satisfied. To do this, we need a 0-based coordinate .tsv file with the following columns: `motif`, `region_id`, `chrom`, `motif_window_start`, `motif_window_end`, `motif_length`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157560, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_length</th>\n",
       "      <th>motif</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>region_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154541</td>\n",
       "      <td>3154555</td>\n",
       "      <td>Oct4-Sox2_0_2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155627</td>\n",
       "      <td>3155641</td>\n",
       "      <td>Oct4-Sox2_1_3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155712</td>\n",
       "      <td>3155726</td>\n",
       "      <td>Oct4-Sox2_2_3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3343669</td>\n",
       "      <td>3343683</td>\n",
       "      <td>Oct4-Sox2_3_6</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3483033</td>\n",
       "      <td>3483047</td>\n",
       "      <td>Oct4-Sox2_4_10</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end            name  score strand  motif_length  \\\n",
       "0  chr1  3154541  3154555   Oct4-Sox2_0_2      0      +            14   \n",
       "1  chr1  3155627  3155641   Oct4-Sox2_1_3      0      +            14   \n",
       "2  chr1  3155712  3155726   Oct4-Sox2_2_3      0      +            14   \n",
       "3  chr1  3343669  3343683   Oct4-Sox2_3_6      0      +            14   \n",
       "4  chr1  3483033  3483047  Oct4-Sox2_4_10      0      +            14   \n",
       "\n",
       "       motif motif_id region_id  \n",
       "0  Oct4-Sox2        0         2  \n",
       "1  Oct4-Sox2        1         3  \n",
       "2  Oct4-Sox2        2         3  \n",
       "3  Oct4-Sox2        3         6  \n",
       "4  Oct4-Sox2        4        10  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import set of motifs\n",
    "motifs_df = BedTool(xiong_motifs_path).to_dataframe()\n",
    "motifs_df.columns = ['chrom','start','end','name','score','strand']\n",
    "\n",
    "#Separate motif name\n",
    "motifs_df['motif_length'] = motifs_df['end'] - motifs_df['start']\n",
    "motifs_df['motif'] = [n.split('_')[0] for n in motifs_df.name]\n",
    "motifs_df['motif_id'] = [n.split('_')[1] for n in motifs_df.name]\n",
    "motifs_df['region_id'] = [n.split('_')[2] for n in motifs_df.name]\n",
    "print(motifs_df.shape)\n",
    "\n",
    "motifs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format regions in order to mark the internal motif window coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_chrom</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>region_id</th>\n",
       "      <th>region_score</th>\n",
       "      <th>region_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3034865</td>\n",
       "      <td>3036897</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3061961</td>\n",
       "      <td>3063993</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3153593</td>\n",
       "      <td>3155625</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154660</td>\n",
       "      <td>3156692</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154899</td>\n",
       "      <td>3156931</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  region_chrom  region_start  region_end region_id  region_score region_strand\n",
       "0         chr1       3034865     3036897         0             0             .\n",
       "1         chr1       3061961     3063993         1             0             .\n",
       "2         chr1       3153593     3155625         2             0             .\n",
       "3         chr1       3154660     3156692         3             0             .\n",
       "4         chr1       3154899     3156931         4             0             ."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_df = BedTool(xiong_regions_path).to_dataframe()\n",
    "regions_df.columns = ['region_chrom','region_start','region_end','region_id','region_score','region_strand']\n",
    "regions_df['region_id']=regions_df['region_id'].astype(str)\n",
    "regions_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Match grouped regions to motifs and collect motif position within the windows to obtain `motif_window_start` and `motif_window_end`. We do this now because it is far easier to work with 0-based coordinated in python than in R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs_df = motifs_df.merge(regions_df[['region_start','region_end','region_id']], on = 'region_id', how = 'left')\n",
    "motifs_df['motif_window_start'] = motifs_df['start']-motifs_df['region_start']\n",
    "motifs_df['motif_window_end'] = motifs_df['end']-motifs_df['region_start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157560, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_length</th>\n",
       "      <th>motif</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>region_id</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>motif_window_start</th>\n",
       "      <th>motif_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3154541</td>\n",
       "      <td>3154555</td>\n",
       "      <td>Oct4-Sox2_0_2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3153593</td>\n",
       "      <td>3155625</td>\n",
       "      <td>948</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155627</td>\n",
       "      <td>3155641</td>\n",
       "      <td>Oct4-Sox2_1_3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3154660</td>\n",
       "      <td>3156692</td>\n",
       "      <td>967</td>\n",
       "      <td>981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3155712</td>\n",
       "      <td>3155726</td>\n",
       "      <td>Oct4-Sox2_2_3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3154660</td>\n",
       "      <td>3156692</td>\n",
       "      <td>1052</td>\n",
       "      <td>1066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3343669</td>\n",
       "      <td>3343683</td>\n",
       "      <td>Oct4-Sox2_3_6</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>3342660</td>\n",
       "      <td>3344692</td>\n",
       "      <td>1009</td>\n",
       "      <td>1023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3483033</td>\n",
       "      <td>3483047</td>\n",
       "      <td>Oct4-Sox2_4_10</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>3482000</td>\n",
       "      <td>3484032</td>\n",
       "      <td>1033</td>\n",
       "      <td>1047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3549773</td>\n",
       "      <td>3549787</td>\n",
       "      <td>Oct4-Sox2_5_12</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>3548761</td>\n",
       "      <td>3550793</td>\n",
       "      <td>1012</td>\n",
       "      <td>1026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3681703</td>\n",
       "      <td>3681717</td>\n",
       "      <td>Oct4-Sox2_6_16</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>3680703</td>\n",
       "      <td>3682735</td>\n",
       "      <td>1000</td>\n",
       "      <td>1014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3977384</td>\n",
       "      <td>3977398</td>\n",
       "      <td>Oct4-Sox2_7_21</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>3976282</td>\n",
       "      <td>3978314</td>\n",
       "      <td>1102</td>\n",
       "      <td>1116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3982153</td>\n",
       "      <td>3982167</td>\n",
       "      <td>Oct4-Sox2_8_23</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>3981205</td>\n",
       "      <td>3983237</td>\n",
       "      <td>948</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4214144</td>\n",
       "      <td>4214158</td>\n",
       "      <td>Oct4-Sox2_9_29</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>Oct4-Sox2</td>\n",
       "      <td>9</td>\n",
       "      <td>29</td>\n",
       "      <td>4213137</td>\n",
       "      <td>4215169</td>\n",
       "      <td>1007</td>\n",
       "      <td>1021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end            name  score strand  motif_length  \\\n",
       "0  chr1  3154541  3154555   Oct4-Sox2_0_2      0      +            14   \n",
       "1  chr1  3155627  3155641   Oct4-Sox2_1_3      0      +            14   \n",
       "2  chr1  3155712  3155726   Oct4-Sox2_2_3      0      +            14   \n",
       "3  chr1  3343669  3343683   Oct4-Sox2_3_6      0      +            14   \n",
       "4  chr1  3483033  3483047  Oct4-Sox2_4_10      0      +            14   \n",
       "5  chr1  3549773  3549787  Oct4-Sox2_5_12      0      -            14   \n",
       "6  chr1  3681703  3681717  Oct4-Sox2_6_16      0      +            14   \n",
       "7  chr1  3977384  3977398  Oct4-Sox2_7_21      0      -            14   \n",
       "8  chr1  3982153  3982167  Oct4-Sox2_8_23      0      +            14   \n",
       "9  chr1  4214144  4214158  Oct4-Sox2_9_29      0      +            14   \n",
       "\n",
       "       motif motif_id region_id  region_start  region_end  motif_window_start  \\\n",
       "0  Oct4-Sox2        0         2       3153593     3155625                 948   \n",
       "1  Oct4-Sox2        1         3       3154660     3156692                 967   \n",
       "2  Oct4-Sox2        2         3       3154660     3156692                1052   \n",
       "3  Oct4-Sox2        3         6       3342660     3344692                1009   \n",
       "4  Oct4-Sox2        4        10       3482000     3484032                1033   \n",
       "5  Oct4-Sox2        5        12       3548761     3550793                1012   \n",
       "6  Oct4-Sox2        6        16       3680703     3682735                1000   \n",
       "7  Oct4-Sox2        7        21       3976282     3978314                1102   \n",
       "8  Oct4-Sox2        8        23       3981205     3983237                 948   \n",
       "9  Oct4-Sox2        9        29       4213137     4215169                1007   \n",
       "\n",
       "   motif_window_end  \n",
       "0               962  \n",
       "1               981  \n",
       "2              1066  \n",
       "3              1023  \n",
       "4              1047  \n",
       "5              1026  \n",
       "6              1014  \n",
       "7              1116  \n",
       "8               962  \n",
       "9              1021  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(motifs_df.shape)\n",
    "motifs_df.head(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter motifs on edges of the windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157560, 14)\n"
     ]
    }
   ],
   "source": [
    "motifs_df = motifs_df[((motifs_df['start']  - motifs_df['region_start'])>=20) & ((motifs_df['region_end']  - motifs_df['end'])>=20)]\n",
    "print(motifs_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter motifs that are of poor `fann` similarity matching to seqlet cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157560, 27)\n"
     ]
    }
   ],
   "source": [
    "motifs_w_metadata_df = pd.read_csv(xiong_motifs_w_metadata_path, sep = '\\t').drop(columns = ['start','end','motif', 'strand'])\n",
    "motifs_df['motif_id'] = motifs_df.motif_id.astype(int)\n",
    "motifs_df = motifs_df.merge(motifs_w_metadata_df, how = 'left', on = 'motif_id', suffixes=('', '_y'))\n",
    "print(motifs_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif\n",
      "Klf4         85191\n",
      "Sox2         48704\n",
      "Oct4-Sox2    23665\n",
      "Name: count, dtype: int64\n",
      "count\n",
      "1     68944\n",
      "2     22400\n",
      "3      7735\n",
      "4      2834\n",
      "5       989\n",
      "6       352\n",
      "7       140\n",
      "8        61\n",
      "10       29\n",
      "9        22\n",
      "11        8\n",
      "12        5\n",
      "13        2\n",
      "22        2\n",
      "15        2\n",
      "14        1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Save motifs\n",
    "motifs_df.to_csv(f'tsv/genomic/motifs/xiong_instances_formatted_for_genomic_perturbations_0based.tsv.gz', sep = '\\t', index = False)\n",
    "print(motifs_df.motif.value_counts())\n",
    "print(motifs_df.region_id.value_counts().value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  region_chrom  region_start  region_end region_id  region_score region_strand\n",
      "0         chr1       3034865     3036897         0             0             .\n",
      "1         chr1       3061961     3063993         1             0             .\n",
      "2         chr1       3153593     3155625         2             0             .\n",
      "3         chr1       3154660     3156692         3             0             .\n",
      "4         chr1       3154899     3156931         4             0             .\n"
     ]
    }
   ],
   "source": [
    "#Save regions and rename regions_df columns to match more standard bed formatting for accessibility perturbations\n",
    "# regions_df = regions_df.merge(motifs_df[['region_id','island_count']].drop_duplicates(ignore_index = True, keep = 'first'), \n",
    "#                               on = 'region_id', how = 'left')\n",
    "print(regions_df.head())\n",
    "regions_df.columns = ['chrom', 'start', 'end', 'region_id', 'score', 'strand']\n",
    "regions_df.to_csv(f'tsv/genomic/motifs/xiong_regions_formatted_for_genomic_perturbations_0based.tsv.gz', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate binding and accessibility perturbations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Allow combinations of up to 2 and up to 3 mutations. This is promarily used for cooperativity and comparisons among island effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_bpnet_osknz.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_wt.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_0h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_3h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_6h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_9h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_12h.slurm\n",
      "sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_atac_15h.slurm\n"
     ]
    }
   ],
   "source": [
    "for pertubation_name, perturbation_details in modeling_design_dict.items():\n",
    "    write_genomic_perturb_script(motifs_df_path = f'tsv/genomic/motifs/xiong_instances_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 motif_groups_df_path = f'tsv/genomic/motifs/xiong_regions_formatted_for_genomic_perturbations_0based.tsv.gz', \n",
    "                                 model_dir = perturbation_details['model_dir'], \n",
    "                                 fasta_file = genome,\n",
    "                                 tasks_in_model_order = ','.join(perturbation_details['tasks']),\n",
    "                                 output_prefix = f'tsv/genomic/motifs/genomic_perturbs_xiong_whole_window_{pertubation_name}', \n",
    "                                 output_cmd_path = f'scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_{pertubation_name}.slurm', \n",
    "                                 input_seqlen = input_length, \n",
    "                                 output_seqlen = output_length,\n",
    "                                 trials = trials,\n",
    "                                 comb_max = 2, comb_min = 1)\n",
    "    print(f'sbatch scripts/genomic/motifs/genomic_perturbs_xiong_whole_window_{pertubation_name}.slurm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bpreveal_404] *",
   "language": "python",
   "name": "conda-env-bpreveal_404-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
