---
title: 'Analysis of enhancer context'
author: "Melanie (1028-02098-001-001)"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

The goal of this analysis is to investigate the contrast between in silico and in vivo context of a motif. This will involve understanding the independence of a pioneering even, as well as assessing the impacts of CpGs across enhancers as another facet of context.

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools); library(parallel)
library(dplyr); library(data.table); library(patchwork); library(readr); library(testit)

#KNITR Options
setwd("/n/projects/mw2098/publications/2024_weilert_acc/code/2_analysis/")
figure_filepath<-"figures/15_enhancer_context"
options(knitr.figure_dir=figure_filepath, java.parameters = "- Xmx6g")

#Lab sources
source("scripts/r/granges_common.r")
source("scripts/r/metapeak_common.r")
source("scripts/r/knitr_common.r")
source("scripts/r/caching.r")
source("scripts/r/metapeak_functions.r")

#Specific sources
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(ggseqlogo)
source("scripts/r/motif_functions.r")
library(spectral)

#Pre-existing variables
threads <- 5
bsgenome<-BSgenome.Mmusculus.UCSC.mm10
txdb<-TxDb.Mmusculus.UCSC.mm10.knownGene

#model information
tasks<-c('oct4', 'sox2', 'nanog', 'klf4', 'zic3')
motif_to_task.list<-list('Oct4-Sox2' = 'oct4', 'Sox2' = 'sox2', 'Klf4' = 'klf4', 'Zic3' = 'zic3', 'Nanog' = 'nanog')
motif_to_task.df<-data.frame(task = motif_to_task.list %>% unlist, motif = names(motif_to_task.list))
output_length <- 1000
input_length <- 2032
count_threshold<-20

#experimental data information
motifs.path<-'tsv/mapped_motifs/all_instances_curated_0based_w_perturb.tsv.gz' #'tsv/mapped_motifs/all_instances_curated_1based.tsv.gz'
regions.path<-'tsv/mapped_motifs/all_islands_curated_1based_w_experimental_data.tsv.gz'
seqlet_ids.path<-'tsv/mapped_motifs/all_pattern_ids.tsv'

marg.path<-'tsv/insilico/marginalizations/motifs/all_marginalizations.tsv.gz'
atac_dinuc_marg.path<-'tsv/insilico/marginalizations/dinucleotides/marginalizations_atac_wt.tsv.gz'
bind_dinuc_marg.path<-'tsv/insilico/marginalizations/dinucleotides/marginalizations_bpnet_osknz.tsv.gz'
paired_perturbs.path<-'tsv/genomic/all_pairs_curated_0based_w_perturb.tsv.gz'

enhancer_motifs.path<-'tsv/genomic/enhancers/enhancer_motifs.tsv.gz'
atac_contrib.path<-'shap/atac_wt_fold1_atac_counts.bw'

nexus.bw.list<-list(zic3 = list(pos = 'bw/mesc_zic3_nexus_combined_positive.bw',
                                neg = 'bw/mesc_zic3_nexus_combined_negative.bw'),
                    oct4 = list(pos = 'bw/mesc_oct4_nexus_combined_positive.bw',
                                neg = 'bw/mesc_oct4_nexus_combined_negative.bw'),
                    sox2 = list(pos = 'bw/mesc_sox2_nexus_combined_positive.bw',
                                neg = 'bw/mesc_sox2_nexus_combined_negative.bw'),
                    klf4 = list(pos = 'bw/mesc_klf4_nexus_combined_positive.bw',
                                neg = 'bw/mesc_klf4_nexus_combined_negative.bw'),
                    nanog = list(pos = 'bw/mesc_nanog_nexus_combined_positive.bw',
                                 neg = 'bw/mesc_nanog_nexus_combined_negative.bw'))

models.list<-list(bpnet_osknz = c('oct4','sox2','nanog','klf4','zic3'),
                  atac_wt = c('atac'),
                  atac_0h = c('atac'),
                  atac_3h = c('atac'),
                  atac_6h = c('atac'),
                  atac_9h = c('atac'),
                  atac_12h = c('atac'),
                  atac_15h = c('atac'))
```

In this analysis we left off the Oct-only motif because its mostly ERV-derived and distance-dependence soft syntax doesn't make sense in that context. Won't add to the analysis nor refute it.

# Import motifs and their associated data

Read in motifs with perturbation scores. Calculate distance from different values as a baseline to explore context scores. 

```{r}
atac_peaks.gr<-rtracklayer::import('narrowpeak/mesc_atac_peaks.narrowPeak') %>% GenomicRanges::resize(., 1, 'center')

motifs.df<-readr::read_tsv(motifs.path)
motifs.gr<-motifs.df %>%
  makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F, 
                           seqnames.field = 'seqnames', start.field = 'start', end.field = 'end', strand.field = 'strand') %>%
  plyranges::mutate(atac_max = regionWhichMaxs(GenomicRanges::resize(., 1000, 'center'), 'bw/mesc_native_atac_combined_normalized.bw'),
                    distance_from_atac_max = abs(((start - end)/2 + start) - atac_max),
                    distance_from_peak_summit = GenomicRanges::distanceToNearest(., atac_peaks.gr, ignore.strand = T))

motifs.df<-motifs.gr %>% as.data.frame()
regions.df<-readr::read_tsv(regions.path)
```

Read in all mapped motifs without curation.

```{r}
seqlet_ids.df<-readr::read_tsv(seqlet_ids.path) %>% dplyr::filter(fold == 'fold1')
```

# Compare marginalization and perturbation scores

Collect marginalization and perturbation scores side by side.

```{r}
marg.df<-readr::read_tsv(marg.path) %>% as.data.frame
motifs.df<-motifs.df %>% dplyr::left_join(marg.df) %>% as.data.frame

motifs.df<-lapply(names(motif_to_task.list), function(x){
  motif.df<-motifs.df %>% dplyr::filter(motif==x)
  motif.df$bind_marg<-motif.df[[paste0(motif_to_task.list[[x]], '_marg')]]
  motif.df$acc_marg<-motif.df[['atac_wt_marg']]
  
  motif.df$bind_perturb<-log(motif.df[[paste0('bpnet_osknz.', motif_to_task.list[[x]], '.WT.pred_sum')]]) -
    log(motif.df[[paste0('bpnet_osknz.', motif_to_task.list[[x]], '.dA.pred_sum')]])
  motif.df$acc_perturb<-log(motif.df$`atac_wt.atac.WT.pred_sum`) - log(motif.df$`atac_wt.atac.dA.pred_sum`)
  
  return(motif.df)
}) %>% rbindlist()

readr::write_tsv(motifs.df, 'tsv/mapped_motifs/all_instances_curated_1based_with_scores.tsv.gz')
```

## Showcase Akr1cl enhancer

This enhancer has two mapped Oct4-Sox2 motifs.

```{r}
curated_region_id<-3491
curated_coords<-c(65037600, 65037950)
curated_enhancer_boundaries.gr<-GRanges('chr1', IRanges(curated_coords[1], curated_coords[2]))
```

Generate contribution scores of mapped motifs.

```{r}
curated_enhancer.df<-data.frame(
  acc_contrib = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'shap/atac_wt_fold1_atac_counts.bw', 
                                         keep_region_coordinates = T) %>% t(),
  sox2_contrib = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'shap/bpnet_osknz_fold1_sox2_counts.bw', 
                                         keep_region_coordinates = T) %>% t(),
  sox2_pos_obs = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'bw/mesc_sox2_nexus_combined_positive.bw', 
                                         keep_region_coordinates = T) %>% t(),
  sox2_neg_obs = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'bw/mesc_sox2_nexus_combined_negative.bw', 
                                         keep_region_coordinates = T) %>% t(),  
  sox2_pos_pred = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'preds/bpnet_osknz_fold1_sox2_pos_peaks_all.bw',
                                         keep_region_coordinates = T) %>% t(),
  sox2_neg_pred = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'preds/bpnet_osknz_fold1_sox2_neg_peaks_all.bw',
                                         keep_region_coordinates = T) %>% t(),
  acc_obs = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'bw/mesc_native_atac_cutsites_combined.bw',
                                         keep_region_coordinates = T) %>% t(),
  acc_pred = standard_metapeak_matrix(regions.gr = curated_enhancer_boundaries.gr, sample.cov = 'preds/atac_wt_fold1_combined_atac_atac_peaks_all.bw',
                                         keep_region_coordinates = T) %>% t(),
  
  seq = getSeq(bsgenome, curated_enhancer_boundaries.gr, as.character = T) %>% stringr::str_split_1(pattern = ''),
  position = start(curated_enhancer_boundaries.gr):end(curated_enhancer_boundaries.gr)
)

# Generate sequence logo
acc_contrib.plot<-curated_enhancer.df%>%
  tidyr::pivot_wider(names_from = seq, values_from = acc_contrib) %>% 
  dplyr::select(A, C, G, T) %>% as.matrix() %>% t() %>%
ggseqlogo(., method='custom', seq_type='dna') + 
  scale_y_continuous(name = 'Acc. contribution.')+
  scale_x_discrete(name = 'Genomic position (chr11, bp)')+
  ggtitle('Akr1cl enhancer')
sox2_contrib.plot<-curated_enhancer.df %>%
  tidyr::pivot_wider(names_from = seq, values_from = sox2_contrib) %>% 
  dplyr::select(A, C, G, T) %>% as.matrix() %>% t() %>%
ggseqlogo(., method='custom', seq_type='dna') + 
  scale_y_continuous(name = 'sox2 contribution.')+
  scale_x_discrete(name = 'Genomic position (chr11, bp)')+
  ggtitle('Akr1cl enhancer')
sox2_bind.plot<-curated_enhancer.df %>% 
  ggplot(.)+
  geom_area(aes(x = position, y = sox2_pos_obs))+
  geom_area(aes(x = position, y = sox2_neg_obs * -1))+
  theme_classic()
sox2_pred.plot<-curated_enhancer.df %>% 
  ggplot(.)+
  geom_area(aes(x = position, y = sox2_pos_pred))+
  geom_area(aes(x = position, y = sox2_neg_pred * -1))+
  theme_classic()
atac_obs.plot<-curated_enhancer.df %>% 
  ggplot(.)+
  geom_area(aes(x = position, y = acc_obs))+
  theme_classic()
atac_pred.plot<-curated_enhancer.df %>% 
  ggplot(.)+
  geom_area(aes(x = position, y = acc_pred))+
  theme_classic()


showcase_enhancer.plot<-sox2_bind.plot + sox2_pred.plot + atac_obs.plot + atac_pred.plot + acc_contrib.plot + sox2_contrib.plot + plot_layout(ncol = 1, heights = c(1, 1, 1, 1, 1, 1, 1))
showcase_enhancer.plot
ggsave(paste0(figure_filepath, '/Akr1cl_candidate_contrib.png'), showcase_enhancer.plot, height = 8, width = 10)
ggsave(paste0(figure_filepath, '/Akr1cl_candidate_contrib.pdf'), showcase_enhancer.plot, height = 8, width = 10)
```

## Validate Akr1cl cooperativity

Next, check the pairwise cooperativity of each motif within this enhancer.

```{r}
motifs_of_interest.gr<-motifs.gr %>% subsetByOverlaps(., curated_enhancer_boundaries.gr, ignore.strand = T)
```

Collect cooperativity data across these motifs.

```{r}
models_of_interest<-c('atac_wt', 'bpnet_osknz')
paired_motifs_of_interest.df<-readr::read_tsv(paired_perturbs.path) %>%
  dplyr::filter(`motif_id.x` %in% motifs_of_interest.gr$motif_id,
                `motif_id.y` %in% motifs_of_interest.gr$motif_id) %>%
  as.data.frame()

paired_perturbs.df<-mclapply(models_of_interest, function(x){
  p.df<-lapply(models.list[[x]], function(task){
    paired_motifs_of_interest.df %>% 
      dplyr::select(region_id, pair_name, 
                    motif.x, motif_id.x,seq_match_quantile.x, motif.y, motif_id.y, seq_match_quantile.y,
                    hAB_sum = !!sym(paste0(x, '/', task, '/hAB/pred_sum')),
                    hB_sum = !!sym(paste0(x, '/', task, '/hB/pred_sum')),
                    hA_sum = !!sym(paste0(x, '/', task, '/hA/pred_sum')),
                    null_sum = !!sym(paste0(x, '/', task, '/null/pred_sum')),
                    pattern_center.x, pattern_center.y) %>%
      dplyr::mutate(model_name = x,
                    task = task,
                    add_coop = (hAB_sum - null_sum)/(hA_sum + hB_sum - 2*null_sum),
                    add_coop_effects = (hAB_sum - null_sum) - (hA_sum + hB_sum - 2*null_sum),
                    mult_coop = log(hAB_sum/null_sum)/log((hA_sum*hB_sum)/(null_sum*null_sum)),
                    dir_coop = (hAB_sum - hB_sum + null_sum)/(hA_sum),
                    joint_effects = hAB_sum - null_sum,
                    marginal_effects = hA_sum + hB_sum - 2*null_sum,
                  A_effects = hA_sum - null_sum,
                  B_effects = hB_sum - null_sum,
                distance = abs(pattern_center.x - pattern_center.y))
  }) %>% rbindlist(.)
}, mc.cores = 6) %>% rbindlist(.)
```

Visualize cooperativity and relative trends dictating prediction results.

```{r}
paired_perturbs_labeled.df<-paired_perturbs.df %>% as.data.frame() %>% dplyr::filter(`motif.x` != 'Nanog', `motif.y` != 'Nanog') %>%
  dplyr::filter(task %in% c('atac')) %>%
  dplyr::mutate(motifA_label = paste0(motif.x, '_', motif_id.x),
                motifB_label =  paste0(motif.y, '_', motif_id.y))

#Calculate total distance
distance.df<-paired_perturbs_labeled.df %>% dplyr::group_by(motifA_label) %>% dplyr::summarize(total_distance = sum(distance))
distance.df$motifA_label<-factor(distance.df$motifA_label, levels = rev(distance.df$motifA_label))

coop.df<-paired_perturbs_labeled.df %>% dplyr::group_by(motifA_label) %>% dplyr::summarize(total_add_coop_effects = sum(add_coop_effects))
coop.df$motifA_label<-factor(coop.df$motifA_label, levels = rev(distance.df$motifA_label))

paired_perturbs_labeled.df$motifA_label<-factor(paired_perturbs_labeled.df$motifA_label, levels = rev(distance.df$motifA_label))
paired_perturbs_labeled.df$motifB_label<-factor(paired_perturbs_labeled.df$motifB_label, levels = rev(distance.df$motifA_label))

coop.plot<-(ggplot(paired_perturbs_labeled.df, aes(motifA_label, motifB_label, fill = log(add_coop)))+
  geom_tile()+
  scale_fill_gradient2(high = '#b2182b', mid = 'white', low = '#2166ac', midpoint = 0, name = 'Coop. (log)')+
  facet_grid(. ~ task)+
  theme_classic()) +
  (ggplot(paired_perturbs_labeled.df, aes(motifA_label, motifB_label, fill = distance))+
  geom_tile()+
  scale_fill_gradient2(high = '#88419d', mid = 'white', low = '#2166ac', midpoint = 0, name = 'Coop. (log)')+
  facet_grid(. ~ task)+
  theme_classic()) + 
(ggplot(coop.df, aes(x = motifA_label, y = total_add_coop_effects))+
  geom_bar(stat = 'identity')+
  theme_classic())+
(ggplot(distance.df, aes(x = motifA_label, y = total_distance))+
  geom_bar(stat = 'identity')+
  theme_classic()) + 
  plot_layout(ncol = 1)
coop.plot
ggsave(paste0(figure_filepath, '/Akr1cl_candidate_cooperativity.png'), coop.plot, height = 15, width = 5)
ggsave(paste0(figure_filepath, '/Akr1cl_candidate_cooperativity.pdf'), coop.plot, height = 15, width = 5)
```

# Examine distance to perturbation plots

Plot distance from strongest pioneering motif relative to the context scores of each motif.

```{r}
max_distance<-300
pioneers_per_region.df<-motifs.df %>%
                     dplyr::filter(mapping_state == 'both') %>%
                     dplyr::group_by(region_id) %>%
                     dplyr::slice_max(order_by = acc_perturb, with_ties = F) %>%
                     dplyr::rename(pioneering_center = pattern_center,
                                   pioneering_contrib = acc_perturb,
                                   pioneering_motif = motif) %>%
                     dplyr::select(region_id, pioneering_center, pioneering_contrib, pioneering_motif) %>% unique()

motifs_w_centers.df<-motifs.df %>% dplyr::filter(mapping_state != 'acc') %>%
  dplyr::left_join(regions.df %>% dplyr::select(region_id, CpG_ratio)) %>%
  
  #Find center of most pioneering motif
  dplyr::left_join(pioneers_per_region.df) %>%
  
  # dplyr::filter(motif %in% c('Oct4-Sox2','Sox2', 'Klf4')) %>%
  dplyr::mutate(
    distance_from_pioneering_center_abs = abs(pattern_center - pioneering_center),
    distance_from_pioneering_center_abs_group = plyr::round_any(distance_from_pioneering_center_abs, 10, ceiling),
    distance_from_pioneering_center = pattern_center - pioneering_center,
    seq_match_group = cut(seq_match_quantile, breaks = seq(0, 1, .25), labels = paste0('q', 1:4), include.lowest = T)) %>%
  dplyr::group_by(region_id) %>%
  dplyr::mutate(centrality = sum(distance_from_pioneering_center_abs),
                pioneers_nearby = sum(mapping_state!='acc')) %>%
  dplyr::arrange(pioneering_contrib) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(row = 1:nrow(.)) %>% as.data.frame()
```

Plot with the strongest pioneers depicted as gray.

```{r}
bind_vs_acc_genomic.plot.list<-lapply(names(motif_to_task.list), function(x){
  g<-ggplot(motifs_w_centers.df %>% 
              dplyr::arrange(distance_from_pioneering_center_abs) %>%
              dplyr::filter(motif==x, 
                            distance_from_pioneering_center_abs<=max_distance, 
                            distance_from_pioneering_center_abs!=0), 
            aes(x = bind_perturb, y = acc_perturb))+
    geom_hline(yintercept = 0, linetype = 'dotted')+
    geom_vline(xintercept = 0, linetype = 'dotted')+
    ggrastr::geom_point_rast(data = motifs_w_centers.df %>% 
              dplyr::filter(motif==x, 
                            distance_from_pioneering_center_abs==0) %>%
              dplyr::arrange(distance_from_pioneering_center_abs), aes(x = bind_perturb, y = acc_perturb),
              color = 'gray', size = .1)+
    ggrastr::geom_point_rast(aes(color = distance_from_pioneering_center_abs), size = .1)+
    ggpubr::stat_cor(method = 'spearman', label.x.npc = 'left', label.y.npc = 'top')+
    scale_color_gradientn(colours = c(inferno(5, begin = .3, end = 1)),
                          name = 'Distance from strongest pioneer', 
                          limits = c(0, max_distance))+
    scale_x_continuous(name = 'Binding perturb. log(WT/dA)', limits = c(-1, 4))+
    scale_y_continuous(name = 'Acc perturb. log(WT/dA)', limits = c(-1, 3))+
    ggtitle(x)+
    theme_classic() + theme(legend.position = 'bottom')
  return(g)
})
bind_vs_acc_genomic.plot<-wrap_plots(bind_vs_acc_genomic.plot.list, nrow = 1)
ggsave(paste0(figure_filepath, '/bind_perturb_vs_acc_perturb_vs_pioneering_twophase_scatter.png'), bind_vs_acc_genomic.plot, height = 4, width = 20)
ggsave(paste0(figure_filepath, '/bind_perturb_vs_acc_perturb_vs_pioneering_twophase_scatter.pdf'), bind_vs_acc_genomic.plot, height = 4, width = 20)
```

## Quantify context scores explained by PWMs and proximity

Calculate and report residuals. First do it in a combined model.


```{r}
residuals.df<-rbind(
  lapply(motifs_w_centers.df$motif %>% unique, function(x){
    model<-lm(formula = acc_perturb ~ seq_match_quantile + distance_from_pioneering_center_abs + distance_from_peak_summit.distance + distance_from_atac_max, data =  motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, motif==x))
    af <- anova(model)
    afss <- af$"Sum Sq"
    df<-cbind(af,var_explained=afss/sum(afss)*100) %>%
      dplyr::mutate(motif=x)
    df$comparison<-rownames(df)
    return(df)
  }) %>% rbindlist() %>%
    dplyr::filter(comparison != 'Residuals') %>%
    dplyr::mutate(model = 'acc'),
  
  lapply(motifs_w_centers.df$motif %>% unique, function(x){
    model<-lm(formula = bind_perturb ~ seq_match_quantile + distance_from_pioneering_center_abs + distance_from_peak_summit.distance + distance_from_atac_max, data =  motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, motif==x))
    af <- anova(model)
    afss <- af$"Sum Sq"
    df<-cbind(af,var_explained=afss/sum(afss)*100) %>%
      dplyr::mutate(motif=x)
    df$comparison<-rownames(df)
    return(df)
  }) %>% rbindlist() %>%
    dplyr::filter(comparison != 'Residuals') %>%
    dplyr::mutate(model = 'bind')
)
residuals.df
```

Run single variate regressions instead of multivariate. This may make more sense given the comparisons, so we will do both and make sure that the results don't change the trends we observe.

```{r}
variables<-c('seq_match_quantile', 'distance_from_pioneering_center_abs', 'distance_from_peak_summit.distance', 'distance_from_atac_max')

residuals_indp.df<-rbind(
  lapply(motifs_w_centers.df$motif %>% unique, function(x){
    lapply(variables, function(y){
      motifs_w_centers.df$readout<-motifs_w_centers.df[[y]]
      model<-lm(formula = acc_perturb ~ readout, data =  motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, motif==x))
      af <- anova(model)
      afss <- af$"Sum Sq"
      df<-cbind(af,var_explained=afss/sum(afss)*100) %>%
        dplyr::mutate(motif=x)
      df$comparison<-y
      return(df)
    }) %>% rbindlist()
  }) %>% rbindlist() %>%
    dplyr::filter(!is.na(`F value`)) %>%
    dplyr::mutate(model = 'acc'),
  
  lapply(motifs_w_centers.df$motif %>% unique, function(x){
    lapply(variables, function(y){
      motifs_w_centers.df$readout<-motifs_w_centers.df[[y]]
      model<-lm(formula = bind_perturb ~ readout, data =  motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, motif==x))
      af <- anova(model)
      afss <- af$"Sum Sq"
      df<-cbind(af,var_explained=afss/sum(afss)*100) %>%
        dplyr::mutate(motif=x)
      df$comparison<-y
      return(df)
    }) %>% rbindlist()
  }) %>% rbindlist() %>%
    dplyr::filter(!is.na(`F value`)) %>%
    dplyr::mutate(model = 'bind')
)
residuals_indp.df
```

Compare models run with independence or not (because I'm paranoid and my regression theory is very rusty)

```{r}
residuals.df<-dplyr::left_join(residuals.df %>% dplyr::select(motif, var_explained, comparison, model), 
                 residuals_indp.df %>% dplyr::select(motif, var_explained, comparison, model) %>% dplyr::rename(var_explained_indp = var_explained)) %>%
  dplyr::mutate(diff_var_expl_comb_diff_indp = round(var_explained - var_explained_indp, 2)) %>%
  dplyr::arrange(desc(abs(diff_var_expl_comb_diff_indp)))
```

Yeah, they're similar. Single variate models like distance even more than PWM scores most of the time. Glad that's stable. Correlations also agree.

Visualize residuals

```{r}
plot<-residuals.df %>%
  dplyr::filter(motif %in% c('Oct4-Sox2', 'Sox2', 'Nanog')) %>%
  ggplot(., aes(x = motif, fill = comparison, y = var_explained))+
  geom_bar(stat = 'identity', position = 'dodge')+
  facet_grid(. ~ model)+
  theme_classic() + theme(legend.position = 'bottom')
plot
ggsave(paste0(figure_filepath, '/perturb_vs_pwm_and_dist.png'), plot, height = 2.5, width = 6.5)
ggsave(paste0(figure_filepath, '/perturb_vs_pwm_and_dist.pdf'), plot, height = 2.5, width = 6.5)
```

Provide direct correlations between the two groups

```{r}
#Secondary only
motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, 
                                      distance_from_pioneering_center_abs!=0) %>%
              dplyr::group_by(motif) %>%
  dplyr::summarize(acc_vs_dist = cor(acc_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   acc_vs_pwm = cor(acc_perturb, seq_match_quantile),
                   bind_vs_dist = cor(bind_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   bind_vs_pwm = cor(bind_perturb, seq_match_quantile)) %>% as.data.frame

#All motifs
motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance) %>%
              dplyr::group_by(motif) %>%
  dplyr::summarize(acc_vs_dist = cor(acc_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   acc_vs_atac_max = cor(acc_perturb, distance_from_atac_max, method = 'spearman'), 
                   acc_vs_summit = cor(acc_perturb, distance_from_peak_summit.distance, method = 'spearman'), 
                   acc_vs_pwm = cor(acc_perturb, seq_match_quantile),
                   bind_vs_dist = cor(bind_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   bind_vs_atac_max = cor(bind_perturb, distance_from_atac_max, method = 'spearman'), 
                   bind_vs_summit = cor(bind_perturb, distance_from_peak_summit.distance, method = 'spearman'), 
                   bind_vs_pwm = cor(bind_perturb, seq_match_quantile)
                   ) %>%
  dplyr::filter(motif %in% c('Oct4-Sox2','Sox2', 'Nanog')) %>% as.data.frame

```

## Show Sox2 correlations at specific sequences

Find Sox2 specific-sequences correlations on these plots.

```{r}
motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id)

motifs_w_centers.df %>% 
  dplyr::filter(motif=='Sox2', 
                seq %in% c(motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id, motif=='Sox2') %>% .$seq), 
                distance_from_pioneering_center_abs<=max_distance) %>%
  .$seq %>%
  table()

bind_vs_acc_genomic_vs_dist.plot<-ggplot(motifs_w_centers.df %>% 
                                          dplyr::filter(motif=='Sox2', 
                                                        seq %in% c(motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id, motif=='Sox2') %>% .$seq), 
                                                        distance_from_pioneering_center_abs<=max_distance,
                                                        distance_from_pioneering_center_abs!=0) %>%
                                          dplyr::arrange(distance_from_pioneering_center_abs), 
                                        aes(x = bind_perturb, y = acc_perturb))+
  geom_hline(yintercept = 0, linetype = 'dotted')+
  geom_vline(xintercept = 0, linetype = 'dotted')+
  ggrastr::geom_point_rast(data = motifs_w_centers.df %>% 
                             dplyr::filter(motif=='Sox2', 
                                           seq %in% c(motifs_w_centers.df %>% 
                                                        dplyr::filter(region_id==curated_region_id, motif=='Sox2') %>% .$seq)) %>%
                             dplyr::filter(distance_from_pioneering_center_abs==0) %>%
                             dplyr::arrange(distance_from_pioneering_center_abs), aes(x = bind_perturb, y = acc_perturb),
                           color = 'gray', size = .1)+
  ggrastr::geom_point_rast(aes(color = distance_from_pioneering_center_abs), size = .1)+
  geom_point(data = motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id, motif=='Sox2'), color = 'black')+
  ggpubr::stat_cor(method = 'spearman', label.x.npc = 'left', label.y.npc = 'top')+
  scale_color_gradientn(colours = c(inferno(5, begin = .3, end = 1)),
                        name = 'Distance from strongest pioneer', 
                        limits = c(0, max_distance))+
  scale_x_continuous(name = 'Binding perturb. log(WT/dA)', limits = c(-1, 4))+
  scale_y_continuous(name = 'Acc perturb. log(WT/dA)', limits = c(-1, 3))+
  facet_grid(. ~ seq)+
  ggtitle('Sox2')+
  theme_classic() + theme(legend.position = 'bottom')

ggsave(paste0(figure_filepath, '/bind_perturb_vs_acc_perturb_vs_pioneering_center_scatter_sox2.png'), bind_vs_acc_genomic_vs_dist.plot, height = 4, width = 6.5)
ggsave(paste0(figure_filepath, '/bind_perturb_vs_acc_perturb_vs_pioneering_center_scatter_sox2.pdf'), bind_vs_acc_genomic_vs_dist.plot, height = 4, width = 6.5)

```

Calculate and report correlation values

```{r}
#Secondary only
motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance, 
                                      distance_from_pioneering_center_abs!=0,
                                      motif=='Sox2', 
                                      seq %in% c(motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id, motif=='Sox2') %>% .$seq)) %>%
  dplyr::group_by(motif) %>%
  dplyr::summarize(acc_vs_dist = cor(acc_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   bind_vs_dist = cor(bind_perturb, distance_from_pioneering_center_abs, method = 'spearman')) %>% as.data.frame

#All motifs
motifs_w_centers.df %>% dplyr::filter(distance_from_pioneering_center_abs<=max_distance,
                                      motif=='Sox2', 
                                      seq %in% c(motifs_w_centers.df %>% dplyr::filter(region_id==curated_region_id, motif=='Sox2') %>% .$seq)) %>%
  dplyr::group_by(motif, seq) %>%
  dplyr::summarize(acc_vs_dist = cor(acc_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   acc_vs_atac_max = cor(acc_perturb, distance_from_atac_max, method = 'spearman'), 
                   acc_vs_summit = cor(acc_perturb, distance_from_peak_summit.distance, method = 'spearman'), 
                   acc_vs_pwm = cor(acc_perturb, seq_match_quantile),
                   bind_vs_dist = cor(bind_perturb, distance_from_pioneering_center_abs, method = 'spearman'), 
                   bind_vs_atac_max = cor(bind_perturb, distance_from_atac_max, method = 'spearman'), 
                   bind_vs_summit = cor(bind_perturb, distance_from_peak_summit.distance, method = 'spearman'), 
                   bind_vs_pwm = cor(bind_perturb, seq_match_quantile)
  ) %>% as.data.frame
```

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```












